---
title: "Homework 1 - answers"
author: "Almog Angel"
date: "01/12/2024"
output: html_document
---

In the following tasks you will use statistics and sequence alignment to analyze real COVID-19 world data.

-   Load packages

```{r}
library(tidyverse)
library(Biostrings)
library(msa)
library(msaR)
library(seqinr)
library(ape)
```

-   Set your working directory

```{r}
setwd("C:/Users/Hen/OneDrive - Technion/intro_to_bioinformatics/HW1")

```

## Task 1 - Data cleaning

-   Read "COVID19_line_list_data.csv" into a variable called "covid_data"

```{r}
covid_data <- read_csv("COVID19_line_list_data.csv")

```

-   Take a look at the data

```{r}
View(covid_data)
```

-   Find a way to remove empty columns:

```{r}
# Identify empty columns
empty_columns <- which(colSums(is.na(covid_data)) == nrow(covid_data))

# Remove empty columns
covid_data <- covid_data[, -empty_columns]

```

-   Find a way to deal with missing values (NAs) in "age" and "gender" columns


```{r}
covid_data <- drop_na(covid_data, age, gender)

```

-   Assuming 'death' is a binary variable (0 for no, 1 for yes) that represent categorical data
-   Convert 'death' to a factor class that contains either "yes" or "no"

```{r}
# assuming date represent date of death
date_rows <- grepl("/", covid_data$death)
covid_data$death[date_rows] <- "1"

# filter out rows not in the convention
covid_data <- covid_data %>%
  filter(death == 0 | death == 1)
#
covid_data$death <- ifelse(covid_data$death == 0, "No", "Yes")


```

## Task 2 - Descriptive statistics analysis

-   Print a descriptive summary of "age"

```{r}
summary(covid_data$age)

```

-   Distribution:
-   
    (1) Generate histogram for "age" distribution and color the bars by "death"
-   Use: position = 'identity', bins = 50, alpha = 0.5

```{r}
ggplot(covid_data, aes(x = age, fill = death)) +
  geom_histogram(position = 'identity', bins = 50, alpha = 0.5) +
  labs(title = "Age Distribution by Death",
       x = "Age",
       y = "Frequency") +
  scale_fill_manual(values = c("No" = "blue", "Yes" = "red")) +
  theme(plot.title = element_text(hjust = 0.5))

```

-   

    (2) Generate bar plot that visualize the number of dead and alive per gender

```{r}
ggplot(covid_data, aes(x = gender, fill = death)) +
  geom_bar(position = 'dodge',  alpha = 0.5) +
  labs(title = "Number of Dead and Alive per Gender",
       x = "Gender",
       y = "Count") +
  scale_fill_manual(values = c("No" = "skyblue", "Yes" = "salmon")) +
  theme(plot.title = element_text(hjust = 0.5))


```

## Task 3 - Statistical Testing

-   Run Chi-squared test to investigate the association between gender and death

```{r}
chi_squared_result <- chisq.test(covid_data$gender, covid_data$death)
print(chi_squared_result)

```

-   Why do we use Chi-squared and not t-test?

```         
answer:
While T-test assume that the data is approximately normally distributed, Chi-squared test are non-parametric and do not assume a normal distribution of the data.
It's important to note that the normal distribution is typically associated with continuous variables.
In our case, gender is categorical variable. Therefore, using a normal distribution to model the association between gender and death directly may not be appropriate and we would prefer to use Chi-squared test.
```

-   Define the Null Hypothesis of the Chi-squared Test in this example

```         
answer:
--- The Null Hypothesis (H0) --- 
                                  There is no association between gender and death; the variables are independent.
```

-   What is the p-value? Is it significant (alpha = 0.05)? Write your conclusions from the statistical test results.

```         
answer:
The p-value is 0.005669, which is significantly lower than 0.05 (5%) our alpha value.
Therefore, we would reject the null hypothesis, and conclude that there is a significant association between death and age.
```

According to the media, older people are more likely to die from COVID-19. - Test this claim statistically - Frame your null and alternative hypothesis here:

``` 
--- My null hypothesis --- :
                                    Older people are as likely to die from COVID-19 as younger people.
                                    
--- My alternative hypothesis ---:
                                    Older people are more likely to die from COVID-19.
```

-   Check if the assumption of normality is valid for your data (use alpha of 0.05)

```{r}
death_subset <- covid_data[covid_data$death == "Yes", ]
alive_subset <- covid_data[covid_data$death == "No", ]
shapiro_test_results = c(shapiro.test(death_subset$age)$p.value , shapiro.test(alive_subset$age)$p.value)

```

-   Run a statistical test:
-   If you assume normality -\> use t-test
-   else -\> use Wilcoxon rank-sum test

```{r}
if (all(shapiro_test_results < 0.05)) {
  # Run t test
  test_results <- t.test(death_subset$age, alive_subset$age)
} else {
  # Run Wilcoxon rank-sum test
  test_results <- wilcox.test(death_subset$age, alive_subset$age)
}

print(test_results)

```

-   Report the p-value. Is it significant? Write your conclusions from the statistical test results.

```         
answer:
The p-value is 2.2e-16, which is significantly lower than 0.05 (5%) our alpha value.
Therefore, we reject the null hypothesis - i.e. we conclude that older people are more likely to die from COVID-19.

```

-   Read "owid-covid-data.csv" from "<https://covid.ourworldindata.org/data/owid-covid-data.csv>" into a variable called "covid_world"

```{r}
covid_world <- read_csv("https://covid.ourworldindata.org/data/owid-covid-data.csv")

```

-   Take a look at the data

```{r}
View(covid_world)
```

Your friend suggest that countries with high Human Development Index (HDI) have a greater risk to die from COVID-19

-   Make a second table and call it "covid_hdi_vs_deaths":

    (1) Select rows with "date" equal to "2024-01-21"
    (2) Select the following columns: "location", "human_development_index", "population", "total_deaths"
    (3) Remove rows with NAs
    (4) Calculate the total deaths per million people to a column named "total_deaths_per_million" (round the number)

    ```{r}
   ## (1), (2), (3)
selected_col <- c("location", "human_development_index", "population", "total_deaths")
covid_hdi_vs_deaths <- covid_world[covid_world$date == "2024-01-21", selected_col]
covid_hdi_vs_deaths <- drop_na(covid_hdi_vs_deaths)
  ## (4)
covid_hdi_vs_deaths <- mutate(covid_hdi_vs_deaths, total_deaths_per_million = round((total_deaths/population) * 1e6, 2))
print(covid_hdi_vs_deaths)
    ```

-   Use a scatterplot to visualize the relations between HDI and the total deaths per million

```{r}

ggplot(covid_hdi_vs_deaths, aes(x = human_development_index, y = total_deaths_per_million)) +
  geom_point() +  # Add points
  labs(x = "HDI", y = "Total Deaths per Million",  
       title = "Relationship between HDI and total Deaths per million") +
  theme(plot.title = element_text(hjust = 0.5))


```

-   Calculate the Pearson and Spearman correlation between HDI and the total deaths per million

```{r}

Pearson_cor <- cor(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$total_deaths_per_million, method = "pearson")
cat("Pearson  correlation coefficient:", round(Pearson_cor, 3), "\n")

Spearman_Cor <- cor(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$total_deaths_per_million, method = "spearman")
cat("Spearman correlation coefficient:", round(Spearman_Cor, 3), "\n")

```

-   Report the Spearman correlation coefficient, the strength of the correlation (weak/moderate/strong/very strong) and the direction (negative/positive)

```         
answer:
The Spearman correlation coefficient we got : rho = 0.752, indicates a strong monotonic relationship (rho > 0.7) and with a positive direction (rho > 0).

```

-   Do you agree with the claim that increased in HDI causes high COVID-19 mortality? (write in detail)

```         
answer:
We don't agree with the claim, as the key word is "cause" - i.e the obtained correlation value indicates on some correlation between the two - yet, higher HDI, for example, can indicate in most cases on good healthcare facilities availbale for the population -> which should decrease the mortality caused by COVID-19 (and not increase it). 

Another aspect is the accuracy and relevancy of the data - in countries with low HDI reports (including mortality reports) might be less accurate and up to date, in comparison to countries with high HDI.

Furthermore, we generally know that correlation does not reveal the cause-effect direction - so we can equally assume that high mortality causes higher HDI, but it doesn't sound reasonable that higher mortality rates out of COVID-19 causes lower standard of living.
```

-   Find another feature in "covid_world" that explain the correlation between HDI and COVID-19 mortality
-   Hint: you already have the answer in the first part of the task

```{r}
modifier_featutre <- "median_age"
selected_col <- c("location", "human_development_index", "population", "total_deaths", modifier_featutre)
covid_hdi_vs_deaths <- covid_world[covid_world$date == "2024-01-21", selected_col]
covid_hdi_vs_deaths_age_filter <- covid_hdi_vs_deaths[covid_hdi_vs_deaths$median_age < 24, selected_col]
covid_hdi_vs_deaths_age_filter <- drop_na(covid_hdi_vs_deaths_age_filter)
#
covid_hdi_vs_deaths_age_filter <- mutate(covid_hdi_vs_deaths_age_filter, total_deaths_per_million = round((total_deaths/population) * 1e6, 2))
#
Spearman_Cor <- cor(covid_hdi_vs_deaths_age_filter$human_development_index, covid_hdi_vs_deaths_age_filter$total_deaths_per_million, method = "spearman")
cat("Spearman correlation coefficient:", round(Spearman_Cor, 3), "\n")

```

```         
Write your explanation here:
Now we used the "median_age" column from the original data, to filter out all the locations with median age above 24.
We saw that the obtained correlation is lower than the previous one (0.629 in comparison to 0.752).
This may suggests that age demographics play a role as a modifier featutre of the correlation between HDI and COVID-19 mortality.
The above fits the results we had in the first part, where we saw a strong correlation between age and COVID-19 mortality.
```

## Task 4 - Sequence Alignment

-   Take a look at the "Biostrings" package vignettes

```{r}
browseVignettes("Biostrings")
```

-   Read the file "covid_spike_variants.fasta" This file contain the amino acids sequence of the COVID-19 spike protein from different variants

-   Read the file using the correct function from the package "Biostrings" and assign to a variable called "variants"

```{r}
variants <- readAAStringSet("covid_spike_variants.fasta")
variants

```

-   How many amino acids are in the Alpha variant?

```         
answer:
1270
```

-   Read the documentation for the Multiple Sequence Alignment (msa) function from the package "msa"

```{r}
?msa()
```

-   Run MSA for the sequences in "variants" and assign the results to "variants.msa"

```{r}
variants.msa <- msa(variants)

```

-   Take a look at the results using the package "msaR" (notice: you can scroll right and left to see all the sequence)

```{r}
msaR(AAMultipleAlignment(variants.msa), colorscheme = "taylor")
```

-   

    (1) Which amino-acids appear in position #13? (type "AMINO_ACID_CODE" in the console)

-   

    (2) Write an example of a SNP that can cause the change in amino acid as we see in the Epsilon variant

```         
answer:
(1) Ser, Ile
(2) Single Nucleotide Polymorphisms are variations at a single position in a DNA sequence, and they can lead to changes in amino acids in the corresponding protein. To illustrate a hypothetical SNP that could result in the amino acid difference seen in the Epsilon variant, we can consider the folloowing example:
The original DNA sequence contained the Serine amino acid: AGU, and the variant had a mutation, replacing the G with U, and resulting in the sequence AUU, which is Isoleucine.

```

Phylogenetic tree for the COVID19 variants: - Use the package "seqinr" to generate a distance matrix from the MSA results and save it in a variable called "distMat"

```{r}
distMat <- dist.alignment(msaConvert(variants.msa, type="seqinr::alignment"), matrix = "similarity")
distMat

```

-   Use the package "ape" to apply "neighbor-joining" (nj) clustering algorithm to construct a phylogenetic tree
-   Save the results in a variable called "ptree"

```{r}
ptree <- nj(distMat)
```

-   Look at the phylogenetic tree and answer the following questions:

```{r}
ape::plot.phylo(ptree)
```

-   

    (1) Which variant is the closest to Gamma?

-   

    (2) Which pair seems to emerged from the Alpha variant?
        (a) Epsilon and Delta
        (b) Beta and Gamma
        (c) Epsilon and Delta

```         
answer:
(1) The Beta variant (Beta SoutAfrica Aug20)
(2) (b) Beta and Gamma
```

-   Save this R Markdown as HTML before submitting

(1) Go to "Knit" \> "Knit to HTML"
(2) Change the file name to your IDs (i.e, "123456789_123456789.html")
