---
title: "Bioinformatics Final Project"
author: "Amit Gabay & Hen Israel"
date: "2024-05-23"
output: html_document
---

Abstract
Lung cancer remains one of the leading causes of cancer-related mortality worldwide.
According to the majority of studies, cigarette smoking is recognized as a major risk factor for lung-cancer development. 
Even so, despite the well-established association between smoking and lung cancer, we have found that there is still a gap in our understanding of the molecular mechanisms underlying the relationship between smoking cessation and lung cancer pathogenesis. 
This study aimes to investigate the specific molecular alterations occurring in lung tissue following smoking cessation and their impact on the initiation, progression, and prognosis of smoking-related lung cancer. Using transcriptomic profiling techniques, we analyze gene expression patterns in lung tissue samples obtained from individuals with varying smoking histories, including current smokers, former smokers who had quit smoking, and never-smokers.


Our analysis revealed significant differences in gene expression profiles between individuals who had quit smoking and those who continued to smoke, particularly in genes associated with cellular signaling pathways, DNA repair mechanisms, and immune response. Furthermore, we identified specific molecular signatures associated with smoking cessation that were indicative of a favorable prognosis in individuals with smoking-related lung cancer. These findings provide novel insights into the molecular mechanisms underlying the beneficial effects of smoking cessation on lung cancer pathogenesis and have important implications for personalized prevention, diagnosis, and treatment strategies. By elucidating the molecular underpinnings of smoking-related lung cancer, this study contributes to our understanding of the disease and offers potential avenues for improving clinical outcomes and reducing the burden of lung cancer on individuals and society.



Install required missing packages
```{r}
# install BiocManager
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

#install packages
packages <- c("tidyverse","hrbrthemes", "viridis", "msigdbr", "dplyr", "tibble", "venn","VennDiagram", "grid") 
biocManager_packages <- c("TCGAbiolinks","DESeq2", "GEOquery", "hrbrthemes", "apeglm", "AnnotationDbi", "org.Hs.eg.db","ReportingTools", "EnhancedVolcano", "pheatmap", "clusterProfiler", "fgsea")

not_installed <-packages[!(packages %in% installed.packages()[ , "Package"])]
biocManager_not_installed <- biocManager_packages[!(biocManager_packages %in% installed.packages()[ , "Package"])]

for (package in not_installed) {
  install.packages(package) 
}

for (package in biocManager_not_installed) {
  BiocManager::install(package) 
}

```

import all packages into R
```{r}
library(TCGAbiolinks)
library(tidyverse)
library(DESeq2)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(tidyverse)
library(GEOquery)
library(apeglm)
library(pheatmap)
library(msigdbr)
library(dplyr)
library(tibble)
library(clusterProfiler)
library(ReportingTools)
library(fgsea)
library(hrbrthemes) # ggplot2 themes
library(viridis) # color palettes
library(venn)
library(VennDiagram)
library(grid)
```

### Fetching Lung Cancer CPTAC-3 Project data from GDC

Set working dir
```{r}
setwd("C:/Users/Amit_g/OneDrive - Technion/Documents/Bioinformatics_Final_Project")
```

- Set a query to access TCGA's lung cancer CPTAC-3 project transcriptome data
```{r}
cptac.query <- GDCquery(project = "CPTAC-3", data.category = "Transcriptome Profiling")
```

- view query's results
```{r}
query.res <- getResults(cptac.query)
View(query.res)
#write.csv(query.res, "query_res.csv")
```

- The transcriptome might contain other types of data except mRNA expression
```{r}
table(query.res$data_type)
```

- A standardized preprocessing step for gene expression data analysis includes aligning the raw sequencing reads to the reference genome with a program called STAR to create the counts matrix.
```{r}
table(query.res$analysis_workflow_type)
```

- Note! There are also some normal and other types of samples
```{r}
table(query.res$sample_type)
```

- Fix query and Remove Duplicates 
```{r}
cptac.query <- GDCquery(
  project = "CPTAC-3", 
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"))

cptac.query.2 <- cptac.query
tmp <- cptac.query.2$results[[1]]
tmp <- tmp[!duplicated(tmp$sample.submitter_id), ]
cptac.query.2$results[[1]] <- tmp
rm(tmp)
```
We are aware of the other sample types beside "Primary Tumor" and "Solid Tissue Normal", but when trying to run line 112 with all the samples types, we encountered an error:
"Error in checkBarcodeDefinition(sample.type) : 
  Primary Tumor;Primary Tumor;Primary Tumor was not found. Please select a difinition from the table above"
We got this error for every other sample type. 
After debugging for a while we have decided we have more than enough samples in both types.

- Take a look at the results
```{r}
cptac.query.2$results[[1]]
```

- Download the data
```{r}
GDCdownload(cptac.query.2, directory = "C:/Users/Amit_g/OneDrive - Technion/Documents/Bioinformatics_Final_Project")
```

- Prepare data
```{r}
cptac.data <- GDCprepare(cptac.query.2, directory = "C:/Users/Amit_g/OneDrive - Technion/Documents/Bioinformatics_Final_Project")
```

- Explore results
```{r}
temp <- cptac.data
str(temp)
head(temp)
summary(temp)
```

```{r}
table(colData(cptac.data)$tobacco_smoking_status, useNA = "always")
```

```{r}
table(colData(cptac.data)$tissue_or_organ_of_origin)
```

- Drop unknown and NAs and filter the data to our needs:
(1) Only samples from lung cancer patients
(2) remove samples with partial/no information about smoking habits: 
    Current Reformed Smoker, Duration Not Specified and Smoking history not documented

```{r}
# Filter data based on our needed criteria

tissue <- c( "Lower lobe, lung",
             "Lung, NOS",
             "Middle lobe, lung",
             "Upper lobe, lung")
statuses <- c( "Current Reformed Smoker for < or = 15 yrs",
               "Current Reformed Smoker for > 15 yrs",
               "Lifelong Non-Smoker",
               "Current Smoker")

filtered_data <- cptac.data[, colData(cptac.data)$tissue_or_organ_of_origin %in% tissue &
                              colData(cptac.data)$tobacco_smoking_status %in% statuses]
filtered_data <- filtered_data[, !is.na(colData(filtered_data)$tobacco_smoking_status)]

# Verify the filtered data
head(filtered_data)

nsclc.data <-filtered_data
rm(filtered_data)
rm(cptac.data)
```
- Verify the filtered results
```{r}
table(colData(nsclc.data)$tobacco_smoking_status, useNA = "always")
table(colData(nsclc.data)$tissue_or_organ_of_origin)
```

Lets take a look on the tumor stages:
```{r}
table(colData(nsclc.data)$ajcc_pathologic_stage, useNA = "always")
```

The tumor_stage variable that TCGA provides for this tumor contains both stages and sub-stages, eg stage IIIA or stage IVB. We want to join together the sub-stages, to increase the group size and reduce complexity (and thus increase the power of the logrank statistics).
```{r}
# remove any of the numbers "1", "2" or "3", but only if they are at the end of the name
# eg "Stage IA2" would become simply "Stage IA"
colData(nsclc.data)$ajcc_pathologic_stage = gsub("[123]$", "", colData(nsclc.data)$ajcc_pathologic_stage)

# remove any of the letters "A", "B" or "C", but only if they are at the end of the name
# eg "stage IIIA" would become simply "stage III"
colData(nsclc.data)$ajcc_pathologic_stage = gsub("[ABC]$", "", colData(nsclc.data)$ajcc_pathologic_stage)

# verify the results as expected
table(colData(nsclc.data)$ajcc_pathologic_stage)
```
Continue with that road, lets examine the following fields:
1) cigarettes per day
2) years smoked
3) packs per year

```{r}
selected_columns <- colData(nsclc.data)[, c('tobacco_smoking_status','cigarettes_per_day','pack_years_smoked', 'years_smoked')]
# Display the table using DT
as.data.frame(selected_columns)
```
We can see that for non-smokers, those fields are NAs. We will turn those values to 0 instead for conducting several experiments according to this parameters. 
Furthermore, We will divide them into groups. 

1. for cigarettes per day: those who never smoked in their lives, those who smoked between 1 to 25, and those who smoked more then 25 cigarettes per day.

2. for years smoked: those who never smoked in their lives, those who smoked between 1 to 35, and those who smoked more then 35 years. 

3. for pack years smoked: those who never smoked in their lives, those who smoked between 1 to 40, and those who smoked more then 40 packs.

```{r}
# Extract the column data and convert to data.frame
col_data <- as.data.frame(colData(nsclc.data))
```
```{r}
# Update for non-smokers cigarettes_per_day, years_smoked column, pack_years_smoked to zero
col_data <- col_data %>%
  mutate(cigarettes_per_day = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ cigarettes_per_day  # Keep the original value otherwise
  ),
    years_smoked = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ years_smoked  # Keep the original value otherwise
    ),
  
    pack_years_smoked = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ pack_years_smoked  # Keep the original value otherwise
    )
  )
```
```{r}
# dividing into groups

# Convert to character type
col_data$cigarettes_per_day <- as.character(col_data$cigarettes_per_day)
col_data$years_smoked <- as.character(col_data$years_smoked)
col_data$pack_years_smoked <- as.character(col_data$pack_years_smoked)

# Use mutate with case_when
col_data <- col_data %>%
  mutate(
    cigarettes_per_day_thr = case_when(
      cigarettes_per_day == "0"   ~ "none",
      cigarettes_per_day <  25    ~ "less than 25",
      cigarettes_per_day >= 25    ~ "more than 25",
      is.na(cigarettes_per_day)  ~ NA_character_
    ),
      years_smoked_thr = case_when(
      years_smoked == "0"   ~ "none",
      years_smoked <  35    ~ "less than 35",
      years_smoked >= 35    ~ "more than 35",
      is.na(years_smoked)  ~ NA_character_
    ),
      pack_years_smoked_thr = case_when(
      pack_years_smoked == "0"   ~ "none",
      pack_years_smoked <  40    ~ "less than 40",
      pack_years_smoked >= 40    ~ "more than 40",
      is.na(pack_years_smoked)  ~ NA_character_
    )
  )

```

```{r}
# Convert back to DFrame
col_data <- as(col_data, "DFrame")

# Assign the modified data back to the colData
colData(nsclc.data) <- col_data
```
```{r}
# verify resuts as expected
table(colData(nsclc.data)$cigarettes_per_day_thr)
table(colData(nsclc.data)$years_smoked_thr)
table(colData(nsclc.data)$pack_years_smoked_thr)
```

```{r}
colData(nsclc.data) <- colData(nsclc.data) %>%
  mutate(cigarettes_per_day = case_when(
    cigarettes_per_day ==    ~ 0,
    cigarettes_per_day <  20  ~ "less then 20",
    cigarettes_per_day >= 20  ~ "more then 20",
    is.na(cigarettes_per_day) ~ NA_character_
  ))
```



We will analyze two different subset: 
1. Solid Tissue Normal 
2. Primary Tumor 
```{r}
# Subset for "Solid Tissue Normal"
nsclc.data.normal <- nsclc.data[, colData(nsclc.data)$sample_type == "Solid Tissue Normal"]

# Subset for "Primary Tumor"
nsclc.data.tumor <- nsclc.data[, colData(nsclc.data)$sample_type == "Primary Tumor"]
```

```{r}
table(colData(nsclc.data.normal)$sample_type)
table(colData(nsclc.data.tumor)$sample_type)
```
# Differential gene expression analysis with with DESeq2 For The lung cancer dataset

- Construct a DESeq2 object
  We want to investigate which genes are different between reformed smokers and non-smokers, between     
  reformed smokers and current smokers, and between reformed smokers for over 15 years and reformed 
  smokers for 15 years or less, so our factor of interest is tobacco smoking duration.
  we will examine each experiment on normal tissues and on primary tumors. 
  (DESeqDataSet will turn assays into counts and set smoking status as a factor, that is why we did not
  do it manually)
  
  ## Solid Tissue Normal 
  ----Add your code-----
  
  ## Primary Tumor  
  ----Add your code-----
