---
title: "Bioinformatics Final Project"
author: "Amit Gabay & Hen Israel"
date: "2024-05-23"
output: html_document
---

Abstract
Lung cancer remains one of the leading causes of cancer-related mortality worldwide.
According to the majority of studies, cigarette smoking is recognized as a major risk factor for lung-cancer development. 
Even so, despite the well-established association between smoking and lung cancer, we have found that there is still a gap in our understanding of the molecular mechanisms underlying the relationship between smoking cessation and lung cancer pathogenesis. 
This study aimes to investigate the specific molecular alterations occurring in lung tissue following smoking cessation and their impact on the initiation, progression, and prognosis of smoking-related lung cancer. Using transcriptomic profiling techniques, we analyze gene expression patterns in lung tissue samples obtained from individuals with varying smoking histories, including current smokers, former smokers who had quit smoking, and never-smokers.


Our analysis revealed significant differences in gene expression profiles between individuals who had quit smoking and those who continued to smoke, particularly in genes associated with cellular signaling pathways, DNA repair mechanisms, and immune response. Furthermore, we identified specific molecular signatures associated with smoking cessation that were indicative of a favorable prognosis in individuals with smoking-related lung cancer. These findings provide novel insights into the molecular mechanisms underlying the beneficial effects of smoking cessation on lung cancer pathogenesis and have important implications for personalized prevention, diagnosis, and treatment strategies. By elucidating the molecular underpinnings of smoking-related lung cancer, this study contributes to our understanding of the disease and offers potential avenues for improving clinical outcomes and reducing the burden of lung cancer on individuals and society.



Install required missing packages
```{r}
# install BiocManager
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

#install packages
packages <- c("tidyverse","hrbrthemes", "viridis", "msigdbr", "dplyr", "tibble", "venn","VennDiagram", "grid") 
biocManager_packages <- c("TCGAbiolinks","DESeq2", "GEOquery", "hrbrthemes", "apeglm", "AnnotationDbi", "org.Hs.eg.db","ReportingTools", "EnhancedVolcano", "pheatmap", "clusterProfiler", "fgsea", "survival", "survminer")

not_installed <-packages[!(packages %in% installed.packages()[ , "Package"])]
biocManager_not_installed <- biocManager_packages[!(biocManager_packages %in% installed.packages()[ , "Package"])]

for (package in not_installed) {
  install.packages(package) 
}

for (package in biocManager_not_installed) {
  BiocManager::install(package) 
}

```

import all packages into R
```{r}
library(TCGAbiolinks)
library(tidyverse)
library(DESeq2)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(tidyverse)
library(GEOquery)
library(apeglm)
library(pheatmap)
library(msigdbr)
library(dplyr)
library(tibble)
library(clusterProfiler)
library(ReportingTools)
library(fgsea)
library(hrbrthemes) # ggplot2 themes
library(viridis) # color palettes
library(venn)
library(VennDiagram)
library(grid)
library(survival)
library(survminer)
```

### Fetching Lung Cancer CPTAC-3 Project data from GDC

Set working dir
```{r}
setwd("C:/Users/Amit_g/OneDrive - Technion/Documents/Bioinformatics_Final_Project")
```

- Set a query to access TCGA's lung cancer CPTAC-3 project transcriptome data
```{r}
cptac.query <- GDCquery(project = "CPTAC-3", data.category = "Transcriptome Profiling")
```

- view query's results
```{r}
query.res <- getResults(cptac.query)
View(query.res)
#write.csv(query.res, "query_res.csv")
```

- The transcriptome might contain other types of data except mRNA expression
```{r}
table(query.res$data_type)
```

- A standardized preprocessing step for gene expression data analysis includes aligning the raw sequencing reads to the reference genome with a program called STAR to create the counts matrix.
```{r}
table(query.res$analysis_workflow_type)
```

- Note! There are also some normal and other types of samples
```{r}
table(query.res$sample_type)
```

- Fix query and Remove Duplicates 
```{r}
cptac.query <- GDCquery(
  project = "CPTAC-3", 
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"))

cptac.query.2 <- cptac.query
tmp <- cptac.query.2$results[[1]]
tmp <- tmp[!duplicated(tmp$sample.submitter_id), ]
cptac.query.2$results[[1]] <- tmp
rm(tmp)
```
We are aware of the other sample types beside "Primary Tumor" and "Solid Tissue Normal", but when trying to run line 112 with all the samples types, we encountered an error:
"Error in checkBarcodeDefinition(sample.type) : 
  Primary Tumor;Primary Tumor;Primary Tumor was not found. Please select a difinition from the table above"
We got this error for every other sample type. 
After debugging for a while we have decided we have more than enough samples in both types.

- Take a look at the results
```{r}
cptac.query.2$results[[1]]
```

- Download the data
```{r}
GDCdownload(cptac.query.2, directory = "C:/Users/Amit_g/OneDrive - Technion/Documents/Bioinformatics_Final_Project")
```

- Prepare data
```{r}
cptac.data <- GDCprepare(cptac.query.2, directory = "C:/Users/Amit_g/OneDrive - Technion/Documents/Bioinformatics_Final_Project")
```

- Explore results
```{r}
temp <- cptac.data
str(temp)
head(temp)
summary(temp)
```

```{r}
table(colData(cptac.data)$tobacco_smoking_status, useNA = "always")
```

```{r}
table(colData(cptac.data)$tissue_or_organ_of_origin)
```

- Drop unknown and NAs and filter the data to our needs:
(1) Only samples from lung cancer patients
(2) remove samples with partial/no information about smoking habits: 
    Current Reformed Smoker, Duration Not Specified and Smoking history not documented

```{r}
# Filter data based on our needed criteria

tissue <- c( "Lower lobe, lung",
             "Lung, NOS",
             "Middle lobe, lung",
             "Upper lobe, lung")
statuses <- c( "Current Reformed Smoker for < or = 15 yrs",
               "Current Reformed Smoker for > 15 yrs",
               "Lifelong Non-Smoker",
               "Current Smoker")

filtered_data <- cptac.data[, colData(cptac.data)$tissue_or_organ_of_origin %in% tissue &
                              colData(cptac.data)$tobacco_smoking_status %in% statuses]
filtered_data <- filtered_data[, !is.na(colData(filtered_data)$tobacco_smoking_status)]

# Verify the filtered data
head(filtered_data)

nsclc.data <-filtered_data
rm(filtered_data)
rm(cptac.data)
```
- Verify the filtered results
```{r}
table(colData(nsclc.data)$tobacco_smoking_status, useNA = "always")
table(colData(nsclc.data)$tissue_or_organ_of_origin)
```

Lets take a look on the tumor stages:
```{r}
table(colData(nsclc.data)$ajcc_pathologic_stage, useNA = "always")
```

The tumor_stage variable that TCGA provides for this tumor contains both stages and sub-stages, eg stage IIIA or stage IVB. We want to join together the sub-stages, to increase the group size and reduce complexity (and thus increase the power of the logrank statistics).
```{r}
# remove any of the numbers "1", "2" or "3", but only if they are at the end of the name
# eg "Stage IA2" would become simply "Stage IA"
colData(nsclc.data)$ajcc_pathologic_stage = gsub("[123]$", "", colData(nsclc.data)$ajcc_pathologic_stage)

# remove any of the letters "A", "B" or "C", but only if they are at the end of the name
# eg "stage IIIA" would become simply "stage III"
colData(nsclc.data)$ajcc_pathologic_stage = gsub("[ABC]$", "", colData(nsclc.data)$ajcc_pathologic_stage)

# verify the results as expected
table(colData(nsclc.data)$ajcc_pathologic_stage)
```
Continue with that road, lets examine the following fields:
1) cigarettes per day
2) years smoked
3) packs per year

```{r}
selected_columns <- colData(nsclc.data)[, c('tobacco_smoking_status','cigarettes_per_day','pack_years_smoked', 'years_smoked')]
# Display the table using DT
as.data.frame(selected_columns)
```
We can see that for non-smokers, those fields are NAs. We will turn those values to 0 instead for conducting several experiments according to this parameters. 
Furthermore, We will divide them into groups. 

1. for cigarettes per day: those who never smoked in their lives, those who smoked between 1 to 25, and those who smoked more then 25 cigarettes per day.

2. for years smoked: those who never smoked in their lives, those who smoked between 1 to 35, and those who smoked more then 35 years. 

3. for pack years smoked: those who never smoked in their lives, those who smoked between 1 to 40, and those who smoked more then 40 packs.

```{r}
# Extract the column data and convert to data.frame
col_data <- as.data.frame(colData(nsclc.data))
```
```{r}
# Update for non-smokers cigarettes_per_day, years_smoked column, pack_years_smoked to zero
col_data <- col_data %>%
  mutate(cigarettes_per_day = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ cigarettes_per_day  # Keep the original value otherwise
  ),
    years_smoked = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ years_smoked  # Keep the original value otherwise
    ),
  
    pack_years_smoked = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ pack_years_smoked  # Keep the original value otherwise
    )
  )
```
```{r}
# dividing into groups

# Convert to character type
col_data$cigarettes_per_day <- as.character(col_data$cigarettes_per_day)
col_data$years_smoked <- as.character(col_data$years_smoked)
col_data$pack_years_smoked <- as.character(col_data$pack_years_smoked)

# Use mutate with case_when
col_data <- col_data %>%
  mutate(
    cigarettes_per_day_thr = case_when(
      cigarettes_per_day == "0"   ~ "none",
      cigarettes_per_day <  25    ~ "less than 25",
      cigarettes_per_day >= 25    ~ "more than 25",
      is.na(cigarettes_per_day)  ~ NA_character_
    ),
      years_smoked_thr = case_when(
      years_smoked == "0"   ~ "none",
      years_smoked <  35    ~ "less than 35",
      years_smoked >= 35    ~ "more than 35",
      is.na(years_smoked)  ~ NA_character_
    ),
      pack_years_smoked_thr = case_when(
      pack_years_smoked == "0"   ~ "none",
      pack_years_smoked <  40    ~ "less than 40",
      pack_years_smoked >= 40    ~ "more than 40",
      is.na(pack_years_smoked)  ~ NA_character_
    )
  )

```

```{r}
# Convert back to DFrame
col_data <- as(col_data, "DFrame")

# Assign the modified data back to the colData
colData(nsclc.data) <- col_data
```
```{r}
# verify resuts as expected
table(colData(nsclc.data)$cigarettes_per_day_thr ,useNA = "always")
table(colData(nsclc.data)$years_smoked_thr ,useNA = "always")
table(colData(nsclc.data)$pack_years_smoked_thr ,useNA = "always")
```

We will analyze two different subset: 
1. Solid Tissue Normal 
2. Primary Tumor 
```{r}
# Subset for "Solid Tissue Normal"
nsclc.data.normal <- nsclc.data[, colData(nsclc.data)$sample_type == "Solid Tissue Normal"]

# Subset for "Primary Tumor"
nsclc.data.tumor <- nsclc.data[, colData(nsclc.data)$sample_type == "Primary Tumor"]
```

```{r}
table(colData(nsclc.data.normal)$sample_type)
table(colData(nsclc.data.tumor)$sample_type)
```
# Differential gene expression analysis with with DESeq2 For The lung cancer dataset

- Construct a DESeq2 object
  We want to investigate which genes are different between reformed smokers and non-smokers, between     
  reformed smokers and current smokers, and between reformed smokers for over 15 years and reformed 
  smokers for 15 years or less, so our factor of interest is tobacco smoking duration.
  we will examine each experiment on normal tissues and on primary tumors. 
  (DESeqDataSet will turn assays into counts and set smoking status as a factor, that is why we did not
  do it manually)
  
  ## Solid Tissue Normal 
  ----Add your code-----
  - Perform DE analysis with DESeq2
```{r}
nsclc.normal.dds <- DESeqDataSet(nsclc.data.normal, design = ~tobacco_smoking_status)
nsclc.normal.dds <- DESeq(nsclc.normal.dds)
```

- Take a look at the gene counts data (just to get an idea)
```{r}
normal.counts_matrix <- counts(nsclc.normal.dds)
#View(normal.counts_matrix)
```

### Analyze differential gene expression results

- Getting differential expression results: 
```{r}
res <- results(nsclc.normal.dds)
mcols(res, use.names=T)

summary(res)
```
- distribution of p-values:
```{r}
hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")

plotMA(res)
```
- For shrinking log fold-change we need to know the 'name' of the analysis we 
  want to shrink its results.
```{r}
resultsNames(nsclc.normal.dds)
```
- Now We will filter genes based on the LFC value - Log Fold Change
* LFC = log2 (normalized_counts_group1 / normalized_counts_group2)
  ( Fold Change = the ratio between healthy and sick,  This tells us how many times bigger or smaller
  something has become. For example, if a gene’s activity doubles, that’s a 2-fold change. If it becomes
  half as active, that’s a 0.5-fold change.)
  WE use log to make sure the result is symmetric.
  
-------------------------------------------------------------------------------------------------------
## First We will perform an analysis for Lifelong non-smoker vs.Current reformed smoker for <= 15 years
-------------------------------------------------------------------------------------------------------
```{r}
# Shrink LFC 
resLFC.1 <- lfcShrink(nsclc.normal.dds, coef="tobacco_smoking_status_Lifelong.Non.Smoker_vs_Current.Reformed.Smoker.for...or...15.yrs", type="apeglm")

plotMA(resLFC.1)
```
Our genes are coded in ENSEMBLID, they need to be converted to conventional symbols.
- Adding a column to the results table with the gene's symbols
```{r}
# Map gene symbols to the ENSEMBL gene IDs from our data
resLFC.1$symbol <- mapIds(org.Hs.eg.db,
                        keys=gsub("\\..*", "", rownames(resLFC.1)),
                        column="SYMBOL",
                        keytype="ENSEMBL",
                        multiVals="first") # multiVals: what should mapIds do when there are multiple values that could be returned?
```

- Order the results by pvalue:
```{r}
resOrdered.1 <- resLFC.1[order(resLFC.1$pvalue),]
resOrdered.1
```

- And finally save our results to CSV so we can take a deeper look in excel:
```{r}
write.csv(resOrdered.1, "signif_results_1.csv")
```

### Visualization of gene expression 1

- Individual genes:
Let start by looking at the first gene - SYTL5:
```{r}
i.1 <- which(resOrdered.1$symbol=='SYTL5')
resOrdered.1[i.1,]
```
A DESeq2 function that can let us extract the normalized values of gene LINC01206:
```{r}
d.1 <- plotCounts(nsclc.normal.dds, gene=rownames(resOrdered.1)[i.1], intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c("Lifelong Non-Smoker", "Current Reformed Smoker for < or = 15 yrs")
d.1 <- d.1[d.1$tobacco_smoking_status %in% selected_statuses, ]
d.1
```

- boxplot for the SYTL5 gene
```{r}
ggplot(d.1[d.1$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "SYTL5 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
We can see a slight difference between the counts of those who had never smoked and the ones that
had smoked and stopped.
We tried boxplotting genes 1-20, and the largest difference appeared on gene number 7.

- Lets take a look at the gene at the 7th index
```{r}
resOrdered.1[7,]
```
Typically, a gene is considered significantly differentially expressed if the adjusted p-valueis below a certain threshold, often 0.05. 
In our case, the adjusted p-value for GPR15 is 2.48058e-07, which is much lower than 0.05, indicating that the differential expression is statistically significant.

- Get GPR15 data 
```{r}
d.1 <- plotCounts(nsclc.normal.dds, gene=rownames(resOrdered.1)[7], intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c("Lifelong Non-Smoker", "Current Reformed Smoker for < or = 15 yrs")
d.1 <- d.1[d.1$tobacco_smoking_status %in% selected_statuses, ]
d.1
```
- boxplot for the GPR15 gene 
```{r}
ggplot(d.1[d.1$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "GPR15 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```

- We see that the expression of the GPR15 gene in people who had smoked and stopped is pretty high, while in people who have never smoked it is relatively low (there is a tail of outliers who had never smoked that also have a high expression of this gene, but, as we did in class, we disregard them).
So GPR15 is upregulated in individuals who have quit smoking.

- What is the GPR15 gene?
________________________________________________
 
The higher expression in reformed smokers might reflect an effect of smoking on gene expression, even after cessation, making GPR15 a potential biomarker for the effects of smoking cessation on lung tissue!

- Keep sample names for those with top and bottom quartiles of CEL for survival analysis
```{r}
quartiles.1 <- quantile(d.1[d.1$tobacco_smoking_status == "Lifelong Non-Smoker",]$count, probs = c(0.25, 0.75))
quartiles.1
low.1 <- rownames(d.1[d.1$tobacco_smoking_status == "Lifelong Non-Smoker" & d.1$count <= quartiles.1[1],])
high.1 <- rownames(d.1[d.1$tobacco_smoking_status == "Lifelong Non-Smoker" & d.1$count >= quartiles.1[2],])
```
* We save the lower quartile and the upper quartile, and the samples that correspond to them, for the survival analysis later
```{r}
write.csv(low.1, file = "GPR15.low1.csv", row.names = FALSE)
write.csv(high.1, file = "GPR15.high1.csv", row.names = FALSE)
```

### Visualization of gene expression for multiple genes

- Volcano Plot
```{r}
EnhancedVolcano(resLFC.1,
                lab = resLFC.1$symbol,
                x = 'log2FoldChange',
                y = 'padj',
                labSize=3,
                FCcutoff = 1.0,
                pCutoff = 1e-05)
```
* P-values are naturally very small.
We take Log to increase and plot them on the y-axis, and negate them to make them positive, 
which will be easier to look at.

* Right Side (Positive LFC): Represents genes that are more highly expressed in Reformed Smokers compared to Lifelong Non-Smokers.
Left Side (Negative LFC): Represents genes that are less expressed in Reformed Smokers compared to Lifelong Non-Smokers.

-We are interested in genes that are the most significant (that is, as high as possible on the y-axis), and that have the greatest Fold Change (that is, as far to the left or right as possible).
LFC cutoff is 1, so everything that is to the right of 2 or to the left of -2 is colored differently.
P-value cutoff is 1e-05, so everything above it is also colored differently

We will define our significance thresholds to be:
-log10p >= 5
- -1 <= LFC <=1
```{r}
# Define the significance thresholds
pvalue_threshold <- 5^(-10)
log2fc_threshold <- 1

resLFC_df.1 <- as.data.frame(resLFC.1)

# Filter the data for significant genes
significant_genes.1 <- resLFC_df.1 %>%
  filter(padj < pvalue_threshold & abs(log2FoldChange) > log2fc_threshold)
```

View and save significant genes for farther analysis and deeper look:
```{r}
significant_genes.1
write.csv(significant_genes.1, "significant_genes_1.csv")
```
  ### Observations from Significant Genes in Case 1:

Key significant Genes and Their Potential Roles:
1. ZIC2 ________
2. MT1G ____________
3. SYTL5 _______________ 
4. LOC124900476 _____________

Conclusion: 
____________________
  
  
- We would like to continue and look at the results from some other angles

### We can also visualize multiple genes with a heatmap:
```{r}
# Take top 10 genes with the lowest p-value that are unregulated in Reformed Smokers (log2FoldChange > 0)
selectUp <- resOrdered.1$symbol[resOrdered.1$log2FoldChange > 0][1:10]
# Take top 10 genes with the lowest p-value that are unregulated in Lifelong Non-Smokers (log2FoldChange < 0)
selectDown <- resOrdered.1$symbol[resOrdered.1$log2FoldChange < 0][1:10]
select <- c(selectUp, selectDown)

dds <- nsclc.normal.dds

# Map ENSEMBL IDs to gene symbols
gene_symbols <- mapIds(org.Hs.eg.db,
                       keys = gsub("\\..*", "", rownames(dds)),
                       column = "SYMBOL",
                       keytype = "ENSEMBL")

# Check for missing mappings and adjust if necessary
missing_mappings <- is.na(gene_symbols)
if (any(missing_mappings)) {
  warning(paste(sum(missing_mappings), "ENSEMBL IDs were not mapped to gene symbols."))
  gene_symbols[missing_mappings] <- rownames(dds)[missing_mappings]
}

# Assign the gene symbols to the row names
rownames(dds) <- gene_symbols

# Subset the dataset based on tobacco smoking status
desired_status <- c("Lifelong Non-Smoker", "Current Reformed Smoker for < or = 15 yrs")
subset_samples <- colData(dds)$tobacco_smoking_status %in% desired_status
dds_subset <- dds[, subset_samples]

# Update the annotation data frame
df <- data.frame(row.names = colnames(dds_subset),
                 status = colData(dds_subset)$tobacco_smoking_status)
                  
#nsclc.data$primary_diagnosis
# Ensure selected genes are in the subset
select <- select[select %in% rownames(dds_subset)]

# Get normalized counts
normcounts <- assay(vst(dds_subset, blind = TRUE))

# Plot heatmap
pheatmap(normcounts[select, ],
         cluster_rows = TRUE,
         show_colnames = FALSE,
         cluster_cols = TRUE,
         annotation_col = df,
         scale = 'row',
         cutree_cols = 2,
         cutree_rows = 2)
```
We have a lot of samples so it is a bit hard to see the clusters.

-Let us zoom in: we will cluster the samples based on similarity in gene expression profiles. This will group similar samples together and make it easier to identify patterns.
```{r}
# Perform sample clustering
sample_dist <- dist(t(normcounts))
sample_clusters <- hclust(sample_dist)
sample_order <- sample_clusters$order
```

```{r}
# Select a subset of samples for visualization
num_samples_to_display <- 20
subset_samples <- sample_order[1:num_samples_to_display]

# Plot the heatmap with sample clustering and subsetting
pheatmap(normcounts[select, subset_samples],
         cluster_rows=TRUE,
         show_colnames = FALSE, 
         cluster_cols=TRUE, 
         annotation_col=df, 
         scale = 'row', 
         cutree_cols = 2, 
         cutree_rows = 2)

```
#### Heatmap Observations for Clustered Samples in Case 1:
_________________________________________________________

### To visualized relations between samples we can use PCA
* simplify our complex datasets by reducing their dimensionality while retaining most of the variation present in the dataset

First, lets prepare the data
```{r}
pca_dds.symbol <- nsclc.normal.dds

# Subset the dataset based on tobacco smoking status
desired_status <- c("Lifelong Non-Smoker", "Current Reformed Smoker for < or = 15 yrs")
subset_samples <- colData(pca_dds.symbol)$tobacco_smoking_status %in% desired_status
pca_dds.symbol <- pca_dds.symbol[, subset_samples]
```

lets find 1000 most variable genes:
```{r}
# Normalize the counts
normcounts = assay(vst(pca_dds.symbol, blind=TRUE))

# Calculate the variance per gene and select the top 1000 variable genes
var_per_gene <- apply(normcounts, 1, var)
selectedGenes <- names(var_per_gene[order(var_per_gene, decreasing = TRUE)][1:1000])
normcounts.top1Kvar <- t(normcounts[selectedGenes, ])
```

- Run and plot PCA by smoking status:
```{r}
# Perform PCA
pcaResults = prcomp(normcounts.top1Kvar)
```
```{r}
# Plot PCA results
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$tobacco_smoking_status, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Tobacco Smoking Status") +
  theme_minimal()
```
What we see here??
___________________________________________

- by cigarettes smoked per day:
```{r}
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$cigarettes_per_day_thr, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Cigarettes Per Day") +
  theme_minimal()
```
What we see here??
______________________________________________

- By pathological stage
```{r}
ggplot(data.frame(PC1 = pcaResults$x[,1], PC2 = pcaResults$x[,2], 
                  Stage = colData(pca_dds.symbol)$ajcc_pathologic_stage)) +
  geom_point(aes(x = PC1, y = PC2, color = Stage), size = 2, alpha = 0.6) +
  labs(x = "PC-1", y = "PC-2", color = "AJCC Pathologic Stage") +
  theme_minimal() +
  scale_color_viridis_d()  # Use a discrete color scale from the viridis package
```
What we see here??
______________________________________________

- By Primary Diagnosis:
```{r}
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$primary_diagnosis, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Primary Diagnosis") +
  theme_minimal()
```
What we see here??
______________________________________________

#### PCA results and Conclusions:
______________________________________________

Next, we will perform GSEA to understand the broader biological context and interactions among significant genes.

### GSEA - case 1
- Filter significant genes and remove NA from the first case
```{r}
filter.sgn.genes.1 <- resOrdered.1
filter.sgn.genes.1.nona <- filter.sgn.genes.1[!is.na(filter.sgn.genes.1$padj),]
filter.sgn.genes.1.nona <- filter.sgn.genes.1.nona[filter.sgn.genes.1.nona$padj < 0.05, ]
filter.sgn.genes.1.nona
```
- Next we need to create an ordered vector by the log fold change with the gene 
  symbols as names:
```{r}
# Convert DESeqResults object to a data frame
filter.first_df <- as.data.frame(filter.sgn.genes.1.nona)

# Remove rows with NA in the symbol column
filtered_data_1 <- filter.first_df %>%
  filter(!is.na(symbol))

# Average log2FoldChange for duplicate gene symbols
unique_genes <- filtered_data_1 %>%
  group_by(symbol) %>%
  summarize(log2FoldChange = mean(log2FoldChange, na.rm = TRUE)) %>%
  ungroup()

# Order the unique genes by log2FoldChange in descending order
unique_genes_ordered_1 <- unique_genes %>%
  arrange(desc(log2FoldChange))

# Create a named vector with log2FoldChange values and gene symbols as names
genes_ordered_1 <- setNames(unique_genes_ordered_1$log2FoldChange, unique_genes_ordered_1$symbol)

genes_ordered_1
```
- For the hallmarks pathways gene sets we'll use msigdbr package.
```{r}
# Load hallmark gene sets
hallmarks_1 <- msigdbr(species = "Homo sapiens", category = "H")
hallmarks_list_1 <- split(hallmarks_1$gene_symbol, hallmarks_1$gs_name)
```
- view hallmark data 
```{r}
# Get all genes in hallmarks_list
all_genes_in_hallmarks_1 <- unique(unlist(hallmarks_list_1))

# Get gene identifiers in genes_ordered
genes_in_ordered_1 <- names(genes_ordered_1)

# Check overlap
overlap_genes_1 <- intersect(all_genes_in_hallmarks_1, genes_in_ordered_1)

# Summary of the overlap
cat("Number of genes in hallmarks_list:", length(all_genes_in_hallmarks_1), "\n")
cat("Number of genes in genes_ordered:", length(genes_in_ordered_1), "\n")
cat("Number of overlapping genes:", length(overlap_genes_1), "\n")

```
We received 408  overlapping genes.

- Using GSEA() as done in class
```{r}
# Run GSEA
set.seed(123)  # Set seed for reproducibility
gsea_results_1 <- fgsea(pathways = hallmarks_list_1, 
                      stats = genes_ordered_1, 
                      minSize = 15, 
                      maxSize = 500)
gsea_results_1
# Convert results to a data frame
gsea_results_df_1 <- as.data.frame(gsea_results_1)
# Filter significant results (e.g., padj < 0.05)
significant_results_1 <- gsea_results_df_1 %>% filter(padj < 0.05)
significant_results_1
```
we got 3 pathways!

- Visualizing the results
```{r}
# Create a dot plot of significant pathways
ggplot(significant_results_1, aes(x = reorder(pathway, NES), y = NES)) +
  geom_point(aes(size = -log10(padj), color = NES)) +
  coord_flip() +
  labs(x = "Pathway", y = "Normalized Enrichment Score (NES)", 
       title = "GSEA Results", size = "-log10(padj)", color = "NES") +
  theme_minimal()

```

Results and conclusions:
_____________________________________________________

----------------------------------------------------------------------------------
## Next We will perform a similar analysis for Current Smokers vs. Reformed Smoker
----------------------------------------------------------------------------------

- Current smoker vs. Current reformed smoker for <= 15 years
```{r}
resLFC.2 <- lfcShrink(nsclc.normal.dds, coef="tobacco_smoking_status_Current.Smoker_vs_Current.Reformed.Smoker.for...or...15.yrs", type="apeglm")
resLFC.2$symbol <- mapIds(org.Hs.eg.db,
                        keys=gsub("\\..*", "", rownames(resLFC.2)),
                        column="SYMBOL",
                        keytype="ENSEMBL",
                        multiVals="first") 
```
- Order the results by pvalue:
```{r}
resOrdered.2 <- resLFC.2[order(resLFC.2$pvalue),]
resOrdered.2
```
- And finally save our results to CSV so we can take a deeper look in excel:
```{r}
write.csv(resOrdered.2, "signif_results_2.csv")
```

### Visualization of gene expression 2

- Lets take a look at the first gene CYP1A1 
```{r}
i.2 <- which(resOrdered.2$symbol=='CYP1A1')
resOrdered.2[i.2,]
```
- Extract the normalized values of the gene of CYP1A1:
```{r}
d.2 <- plotCounts(nsclc.normal.dds, gene=rownames(resOrdered.2)[i.2], intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c("Current Smoker", "Current Reformed Smoker for < or = 15 yrs")
d.2 <- d.2[d.2$tobacco_smoking_status %in% selected_statuses, ]
d.2
``` 
-boxplot for CYP1A1:
```{r}
ggplot(d.2[d.2$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "CYP1A1 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size=8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
We can see a difference between the counts of those who still smoking and the ones that
had smoked and stopped.
We tried boxplotting genes 1-20, and the largest difference appeared on gene number 9.
- Lets take a look at the gene at the 16th index
```{r}
resOrdered.2[9,]
```
We can see the adjusted p-value for HES2 is 2.2491e-11, which is much lower than 0.05, indicating that the differential expression is statistically significant.

- Extract the normalized values of the gene of HES2
```{r}
d.2 <- plotCounts(nsclc.normal.dds, gene=rownames(resOrdered.2)[9], intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c("Current Smoker", "Current Reformed Smoker for < or = 15 yrs")
d.2 <- d.2[d.2$tobacco_smoking_status %in% selected_statuses, ]
d.2
``` 
```{r}
ggplot(d.2[d.2$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "HES2 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
- We can see that the expression of the HES2 gene in people who still smoking is higher then those who stopped.
The upregulation of HES2 in current smokers compared to smokers who quit suggests that this gene may play a role in the biological processes associated with smoking cessation. 

- What is the HES2?
_______________________________________________

- Keep sample names for those with top and bottom quartiles of WSCD2 for later
```{r}
quartiles.2 <- quantile(d.2[d.2$tobacco_smoking_status == "Current Reformed Smoker for < or = 15 yrs",]$count, probs = c(0.25, 0.75))
quartiles.2
low.2 <- rownames(d.2[d.2$tobacco_smoking_status == "Current Reformed Smoker for < or = 15 yrs" & d.2$count <= quartiles.2[1],])
high.2 <- rownames(d.2[d.2$tobacco_smoking_status == "Current Reformed Smoker for < or = 15 yrs" & d.2$count >= quartiles.2[2],])
```
* We save the lower quartile and the upper quartile, and the samples that correspond to them, for the survival analysis later
```{r}
write.csv(low.2, file = "HES2.low2.csv", row.names = FALSE)
write.csv(high.2, file = "HES2.high2.csv", row.names = FALSE)
```

### Visualization of gene expression for multiple genes

- Volcano Plot
```{r}
EnhancedVolcano(resLFC.2,
                lab = resLFC.2$symbol,
                x = 'log2FoldChange',
                y = 'padj',
                labSize=3)
```
* Right Side: Genes with positive LFC are upregulated in the group specified first in Current Smokers.
Left Side: Genes with negative LFC are upregulated in the group specified second in Current Reformed Smokers.

We will define our significance thresholds to be the default ones:
-log10p >= 5
-1.5 <= LFC <=1.5
```{r}
# Define the significance thresholds
pvalue_threshold.2 <- 10^(-5)
log2fc_threshold.2 <- 1.5

resLFC_df.2 <- as.data.frame(resLFC.2)

# Filter the data for significant genes
significant_genes.2 <- resLFC_df.2 %>%
  filter(padj < pvalue_threshold.2 & abs(log2FoldChange) > log2fc_threshold.2)
```

View and save significant genes for farther analysis and deeper look:
```{r}
significant_genes.2
write.csv(significant_genes.2, "significant_genes_2.csv")
```
### Observations from Significant Genes in Case 2:

Key Significant Genes and Their Potential Roles:
1. ZIC2 - *also cum up in case 1* 
2. ALDH3A1 _________
3. CYP1A1 ____________
4. MTND1P23 _________

Conclusion:
____________________

### visualize multiple genes with a heatmap:
```{r}
# Take top 10 genes with the lowest p-value that are unregulated in Reformed Smokers (log2FoldChange > 0)
selectUp <- resOrdered.2$symbol[resOrdered.2$log2FoldChange > 0][1:10]
# Take top 10 genes with the lowest p-value that are unregulated in Lifelong Non-Smokers (log2FoldChange < 0)
selectDown <- resOrdered.2$symbol[resOrdered.2$log2FoldChange < 0][1:10]
select <- c(selectUp, selectDown)

dds <- nsclc.normal.dds

# Map ENSEMBL IDs to gene symbols
gene_symbols <- mapIds(org.Hs.eg.db,
                       keys = gsub("\\..*", "", rownames(dds)),
                       column = "SYMBOL",
                       keytype = "ENSEMBL")

# Check for missing mappings and adjust if necessary
missing_mappings <- is.na(gene_symbols)
if (any(missing_mappings)) {
  warning(paste(sum(missing_mappings), "ENSEMBL IDs were not mapped to gene symbols."))
  gene_symbols[missing_mappings] <- rownames(dds)[missing_mappings]
}

# Assign the gene symbols to the row names
rownames(dds) <- gene_symbols

# Subset the dataset based on tobacco smoking status
desired_status <- c("Current Smoker", "Current Reformed Smoker for < or = 15 yrs")
subset_samples <- colData(dds)$tobacco_smoking_status %in% desired_status
dds_subset <- dds[, subset_samples]

# Update the annotation data frame
df <- data.frame(row.names = colnames(dds_subset),
                 status = colData(dds_subset)$tobacco_smoking_status,
                 cigarettes = colData(dds_subset)$cigarettes_per_day_thr)

# Ensure selected genes are in the subset
select <- select[select %in% rownames(dds_subset)]

# Get normalized counts
normcounts <- assay(vst(dds_subset, blind = TRUE))

# Plot heatmap
pheatmap(normcounts[select, ],
         cluster_rows = TRUE,
         show_colnames = FALSE,
         cluster_cols = TRUE,
         annotation_col = df,
         scale = 'row',
         cutree_cols = 2,
         cutree_rows = 2)
```
We have a lot of samples so it is a bit hard to see the clusters.

-Let us zoom in: we will cluster the samples based on similarity in gene expression profiles. This will group similar samples together and make it easier to identify patterns.
```{r}
# Perform sample clustering
sample_dist <- dist(t(normcounts))
sample_clusters <- hclust(sample_dist)
sample_order <- sample_clusters$order
```

```{r}
# Select a subset of samples for visualization
num_samples_to_display <- 20
subset_samples <- sample_order[1:num_samples_to_display]

# Plot the heatmap with sample clustering and subsetting
pheatmap(normcounts[select, subset_samples],
         cluster_rows=TRUE,
         show_colnames = FALSE, 
         cluster_cols=TRUE, 
         annotation_col=df, 
         scale = 'row', 
         cutree_cols = 2, 
         cutree_rows = 2)

```
#### Heatmap Observations for Clustered Samples in Case 1:
_________________________________________________________

### To visualized relations between samples we can use PCA

First, lets find 1000 most variable genes:
```{r}
pca_dds.symbol <- nsclc.normal.dds

# Subset the dataset based on tobacco smoking status
desired_status <- c("Current Smoker", "Current Reformed Smoker for < or = 15 yrs")
subset_samples <- colData(pca_dds.symbol)$tobacco_smoking_status %in% desired_status
pca_dds.symbol <- pca_dds.symbol[, subset_samples]

# Normalize the counts
normcounts = assay(vst(pca_dds.symbol, blind=TRUE))

# Calculate the variance per gene and select the top 1000 variable genes
var_per_gene <- apply(normcounts, 1, var)
selectedGenes <- names(var_per_gene[order(var_per_gene, decreasing = TRUE)][1:1000])
normcounts.top1Kvar <- t(normcounts[selectedGenes, ])
```

- Run and plot PCA by smoking status:
```{r}
# Perform PCA
pcaResults = prcomp(normcounts.top1Kvar)
```
```{r}
# Plot PCA results
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$tobacco_smoking_status, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Tobacco Smoking Status") +
  theme_minimal()
```
What we see here??
___________________________________________

- by cigarettes smoked per day:
```{r}
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$cigarettes_per_day_thr, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Cigarettes Per Day") +
  theme_minimal()
```
What we see here??
______________________________________________

- By pathological stage
```{r}
ggplot(data.frame(PC1 = pcaResults$x[,1], PC2 = pcaResults$x[,2], 
                  Stage = colData(pca_dds.symbol)$ajcc_pathologic_stage)) +
  geom_point(aes(x = PC1, y = PC2, color = Stage), size = 2, alpha = 0.6) +
  labs(x = "PC-1", y = "PC-2", color = "AJCC Pathologic Stage") +
  theme_minimal() +
  scale_color_viridis_d()  # Use a discrete color scale from the viridis package
```
What we see here??
______________________________________________

- By Primary Diagnosis:
```{r}
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$primary_diagnosis, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Primary Diagnosis") +
  theme_minimal()
```
What we see here??
______________________________________________

#### PCA results and Conclusions:
______________________________________________

Next, we will perform GSEA to understand the broader biological context and interactions among significant genes.
### GSEA - case 2

- Filter significant genes and remove NA from the second case
```{r}
filter.sgn.genes.2 <- resOrdered.2
filter.sgn.genes.2.nona <- filter.sgn.genes.2[!is.na(filter.sgn.genes.2$padj),]
filter.sgn.genes.2.nona <- filter.sgn.genes.2.nona[filter.sgn.genes.2.nona$padj < 0.05, ]
filter.sgn.genes.2.nona
```
- Next we need to create an ordered vector by the log fold change with the gene 
  symbols as names:
```{r}
# Convert DESeqResults object to a data frame
filter.second_df <- as.data.frame(filter.sgn.genes.2.nona)

# Remove rows with NA in the symbol column
filtered_data_2 <- filter.second_df %>%
  filter(!is.na(symbol))

# Average log2FoldChange for duplicate gene symbols
unique_genes_2 <- filtered_data_2 %>%
  group_by(symbol) %>%
  summarize(log2FoldChange = mean(log2FoldChange, na.rm = TRUE)) %>%
  ungroup()

# Order the unique genes by log2FoldChange in descending order
unique_genes_ordered_2 <- unique_genes_2 %>%
  arrange(desc(log2FoldChange))

# Create a named vector with log2FoldChange values and gene symbols as names
genes_ordered_2 <- setNames(unique_genes_ordered_2$log2FoldChange, unique_genes_ordered_2$symbol)

genes_ordered_2
```

- For the hallmarks pathways gene sets we'll use msigdbr package.
```{r}
# Load hallmark gene sets
hallmarks_2 <- msigdbr(species = "Homo sapiens", category = "H")
hallmarks_list_2 <- split(hallmarks_2$gene_symbol, hallmarks_2$gs_name)
```
- view hallmark data 
```{r}
# Get all genes in hallmarks_list
all_genes_in_hallmarks_2 <- unique(unlist(hallmarks_list_2))

# Get gene identifiers in genes_ordered
genes_in_ordered_2 <- names(genes_ordered_2)

# Check overlap
overlap_genes_2 <- intersect(all_genes_in_hallmarks_2, genes_in_ordered_2)

# Summary of the overlap
cat("Number of genes in hallmarks_list:", length(all_genes_in_hallmarks_2), "\n")
cat("Number of genes in genes_ordered:", length(genes_in_ordered_2), "\n")
cat("Number of overlapping genes:", length(overlap_genes_2), "\n")
```
We received 914  genes.

- Using GSEA() as done in class
```{r}
# Run GSEA
set.seed(123)  # Set seed for reproducibility
gsea_results_2 <- fgsea(pathways = hallmarks_list_2, 
                      stats = genes_ordered_2, 
                      minSize = 15, 
                      maxSize = 500)
gsea_results_2
# Convert results to a data frame
gsea_results_df_2 <- as.data.frame(gsea_results_2)
# Filter significant results (e.g., padj < 0.05)
significant_results_2 <- gsea_results_df_2 %>% filter(padj < 0.05)
significant_results_2
```
we got 19 pathways!

- Visualizing the results
```{r}
# Create a dot plot of significant pathways
ggplot(significant_results_2, aes(x = reorder(pathway, NES), y = NES)) +
  geom_point(aes(size = -log10(padj), color = NES)) +
  coord_flip() +
  labs(x = "Pathway", y = "Normalized Enrichment Score (NES)", 
       title = "GSEA Results", size = "-log10(padj)", color = "NES") +
  theme_minimal()
```
Results and conclusions:
_____________________________________________________


---------------------------------------------------------------------------------------------
## Last,  We will perform a similar analysis for refoemred smokers of different time durations
---------------------------------------------------------------------------------------------

- Current reformed smoker for <= 15 years vs. Current reformed smoker for > 15 years
```{r}
resLFC.3 <- lfcShrink(nsclc.normal.dds, coef="tobacco_smoking_status_Current.Reformed.Smoker.for...15.yrs_vs_Current.Reformed.Smoker.for...or...15.yrs" , type="apeglm")
resLFC.3$symbol <- mapIds(org.Hs.eg.db,
                        keys=gsub("\\..*", "", rownames(resLFC.3)),
                        column="SYMBOL",
                        keytype="ENSEMBL",
                        multiVals="first") 
```
- Order the results by pvalue:
```{r}
resOrdered.3 <- resLFC.3[order(resLFC.3$pvalue),]
resOrdered.3
```
- And finally save our results to CSV so we can take a deeper look in excel:
```{r}
write.csv(resOrdered.3, "signif_results_3.csv")
```

### Visualization of gene expression 3

Lets take a look at the first gene - ZIC2:
```{r}
i.3 <- which(resOrdered.3$symbol=='ZIC2')
resOrdered.3[i.3,]
```
- Extract the normalized values of the gene of ZIC2:
```{r}
d.3 <- plotCounts(nsclc.normal.dds, gene=rownames(resOrdered.3)[i.3], intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c("Current Reformed Smoker for > 15 yrs", "Current Reformed Smoker for < or = 15 yrs")
d.3 <- d.3[d.3$tobacco_smoking_status %in% selected_statuses, ]
d.3
``` 
- boxplot for the ALB gene
```{r}
ggplot(d.3[d.3$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "ZIC2 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 6),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
``` 
As we mention, high levels of ZIC2 expression are associated with advanced tumor stage and decreased overall survival.
We can may suppose, that people who stopped smoking, may have better survival chance due to the low level of ZIC2, and maybe quitting from cigarettes can regulate the level of ZIC2 expression.
The outilers are probably those who have advanced tumor stage or quit smoking recently but we do not have this information to be sure. 

We tried boxplotting genes 1-20, and the largest difference appeared on gene number 4.
- Lets take a look at the gene at the 4th index
```{r}
resOrdered.3[4,]
```
In our case, the adjusted p-value for LINC01108 is 2.68252e-07, which is much lower than 0.05, indicating that the differential expression is statistically significant.

- extract the normalized values of gene ATP2A1: 
```{r}
resOrdered.3[4,]
d.3 <- plotCounts(nsclc.normal.dds, gene=rownames(resOrdered.3)[4], intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c("Current Reformed Smoker for > 15 yrs", "Current Reformed Smoker for < or = 15 yrs")
d.3 <- d.3[d.3$tobacco_smoking_status %in% selected_statuses, ]
d.3
```

- boxplot for the LINC01108 gene 
```{r}
ggplot(d.3[d.3$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "LINC01108 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 6),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
- We see that the expression of the LINC01108 gene in people who had smoked and stopped for 15 years or less is pretty high, while in people who had smoked and stopped for more than 15 years is relatively low.

LINC01108
______________________________________


Conclusion:
___________________________________

- Keep sample names for those with top and bottom quartiles of LINC01108 for later
```{r}
quartiles.3 <- quantile(d.3[d.3$tobacco_smoking_status == "Current Reformed Smoker for > 15 yrs",]$count, probs = c(0.25, 0.75))
quartiles.3
low.3 <- rownames(d.3[d.3$tobacco_smoking_status == "Current Reformed Smoker for > 15 yrs" & d.3$count <= quartiles.3[1],])
high.3 <- rownames(d.3[d.3$tobacco_smoking_status == "Current Reformed Smoker for > 15 yrs" & d.3$count >= quartiles.3[2],])
```
* We save the lower quartile and the upper quartile, and the samples that correspond to them, for the survival analysis later
```{r}
write.csv(low.3, file = "LINC01108.low3.csv", row.names = FALSE)
write.csv(high.3, file = "LINC01108.high3.csv", row.names = FALSE)
```

## Visualization of gene expression for multiple genes

- Volcano Plot
```{r}
EnhancedVolcano(resLFC.3,
                lab = resLFC.3$symbol,
                x = 'log2FoldChange',
                y = 'padj',
                labSize=3,
                pCutoff=1e-01,
                FCcutoff=0.5)
```
* Right Side: Genes with positive LFC are upregulated in the group of reformed smokers for more than 15 years.
Left Side: Genes with negative LFC are upregulated in the group of reformed smokers for less than or equal to 15 years.

We will define our significance thresholds to be:
-log10p >= 1
-0.5 <= LFC <=0.5
```{r}
# Define the significance thresholds
pvalue_threshold <- 10^(-1)
log2fc_threshold <- 0.5

resLFC_df.3 <- as.data.frame(resLFC.3)

# Filter the data for significant genes
significant_genes.3 <- resLFC_df.3 %>%
  filter(padj < pvalue_threshold & abs(log2FoldChange) > log2fc_threshold)
```
View and save significant genes for farther analysis and deeper look:
```{r}
significant_genes.3
write.csv(significant_genes.3, "significant_genes_3.csv")
```
### Observations from Significant Genes in Case 2:

Key Significant Genes and Their Potential Roles:
1. TEDC2 ___________
2. MYOZ1 _________
3. LOC124900476 ____________

Conclusion:
____________________

### visualize multiple genes with a heatmap:
```{r}
# Take top 10 genes with the lowest p-value that are unregulated in Reformed Smokers (log2FoldChange > 0)
selectUp <- resOrdered.3$symbol[resOrdered.3$log2FoldChange > 0][1:10]
# Take top 10 genes with the lowest p-value that are unregulated in Lifelong Non-Smokers (log2FoldChange < 0)
selectDown <- resOrdered.3$symbol[resOrdered.3$log2FoldChange < 0][1:10]
select <- c(selectUp, selectDown)

dds <- nsclc.normal.dds

# Map ENSEMBL IDs to gene symbols
gene_symbols <- mapIds(org.Hs.eg.db,
                       keys = gsub("\\..*", "", rownames(dds)),
                       column = "SYMBOL",
                       keytype = "ENSEMBL")

# Check for missing mappings and adjust if necessary
missing_mappings <- is.na(gene_symbols)
if (any(missing_mappings)) {
  warning(paste(sum(missing_mappings), "ENSEMBL IDs were not mapped to gene symbols."))
  gene_symbols[missing_mappings] <- rownames(dds)[missing_mappings]
}

# Assign the gene symbols to the row names
rownames(dds) <- gene_symbols

# Subset the dataset based on tobacco smoking status
desired_status <- c("Current Reformed Smoker for > 15 yrs", "Current Reformed Smoker for < or = 15 yrs")
subset_samples <- colData(dds)$tobacco_smoking_status %in% desired_status
dds_subset <- dds[, subset_samples]

# Update the annotation data frame
df <- data.frame(row.names = colnames(dds_subset),
                 status = colData(dds_subset)$tobacco_smoking_status,
                 cigarettes = colData(dds_subset)$cigarettes_per_day_thr)


# Ensure selected genes are in the subset
select <- select[select %in% rownames(dds_subset)]

# Get normalized counts
normcounts <- assay(vst(dds_subset, blind = TRUE))

# Plot heatmap
pheatmap(normcounts[select, ],
         cluster_rows = TRUE,
         show_colnames = FALSE,
         cluster_cols = TRUE,
         annotation_col = df,
         scale = 'row',
         cutree_cols = 2,
         cutree_rows = 2)
```
We have a lot of samples so it is a bit hard to see the clusters.
-Let us zoom in: we will cluster the samples based on similarity in gene expression profiles. This will group similar samples together and make it easier to identify patterns.
```{r}
# Perform sample clustering
sample_dist <- dist(t(normcounts))
sample_clusters <- hclust(sample_dist)
sample_order <- sample_clusters$order
```

```{r}
# Select a subset of samples for visualization
num_samples_to_display <- 20
subset_samples <- sample_order[1:num_samples_to_display]

# Plot the heatmap with sample clustering and subsetting
pheatmap(normcounts[select, subset_samples],
         cluster_rows=TRUE,
         show_colnames = FALSE, 
         cluster_cols=TRUE, 
         annotation_col=df, 
         scale = 'row', 
         cutree_cols = 2, 
         cutree_rows = 2)

```
### Heatmap Observations for Clustered Samples in Case 3:
_________________________________________________________

### PCA to visualized relations between samples
First, lets find 1000 most variable genes:
```{r}
pca_dds.symbol <- nsclc.normal.dds

# Subset the dataset based on tobacco smoking status
desired_status <- c("Current Reformed Smoker for > 15 yrs", "Current Reformed Smoker for < or = 15 yrs")
subset_samples <- colData(pca_dds.symbol)$tobacco_smoking_status %in% desired_status
pca_dds.symbol <- pca_dds.symbol[, subset_samples]

# Normalize the counts
normcounts = assay(vst(pca_dds.symbol, blind=TRUE))

# Calculate the variance per gene and select the top 1000 variable genes
var_per_gene <- apply(normcounts, 1, var)
selectedGenes <- names(var_per_gene[order(var_per_gene, decreasing = TRUE)][1:1000])
normcounts.top1Kvar <- t(normcounts[selectedGenes, ])
```

Run and plot PCA by Smoking Status:
```{r}
# Perform PCA
pcaResults = prcomp(normcounts.top1Kvar)

# Plot PCA results
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$tobacco_smoking_status, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Tobacco Smoking Status") +
  theme_minimal()
```
What we see here??
___________________________________________

- by cigarettes smoked per day:
```{r}
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$cigarettes_per_day_thr, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Cigarettes Per Day") +
  theme_minimal()
```
What we see here??
______________________________________________

- By pathological stage
```{r}
ggplot(data.frame(PC1 = pcaResults$x[,1], PC2 = pcaResults$x[,2], 
                  Stage = colData(pca_dds.symbol)$ajcc_pathologic_stage)) +
  geom_point(aes(x = PC1, y = PC2, color = Stage), size = 2, alpha = 0.6) +
  labs(x = "PC-1", y = "PC-2", color = "AJCC Pathologic Stage") +
  theme_minimal() +
  scale_color_viridis_d()  # Use a discrete color scale from the viridis package
```
What we see here??
______________________________________________

- By Primary Diagnosis:
```{r}
qplot(pcaResults$x[,1], pcaResults$x[,2], 
      col=colData(pca_dds.symbol)$primary_diagnosis, 
      size=I(2), alpha=I(0.6)) +
  labs(x="PC-1", y="PC-2", color="Primary Diagnosis") +
  theme_minimal()
```
What we see here??
______________________________________________

#### PCA results and Conclusions:
______________________________________________

### GSEA - case 3

- Filter significant genes and removes NAs from third case
```{r}
# Filter out non-significant genes from third sace
filter.sgn.genes.3 <- resOrdered.3
filter.sgn.genes.3.nona <- filter.sgn.genes.3[!is.na(filter.sgn.genes.3$padj),]
filter.sgn.genes.3.nona <- filter.sgn.genes.3.nona[filter.sgn.genes.3.nona$padj < 0.05, ]
filter.sgn.genes.3.nona
```
- Next we need to create an ordered vector by the log fold change with the gene 
  symbols as names:
```{r}

# Convert DESeqResults object to a data frame
filter.third_df <- as.data.frame(filter.sgn.genes.3.nona)

# Remove rows with NA in the symbol column
filtered_data_3 <- filter.third_df %>%
  filter(!is.na(symbol))

# Average log2FoldChange for duplicate gene symbols
unique_genes_3 <- filtered_data_3 %>%
  group_by(symbol) %>%
  summarize(log2FoldChange = mean(log2FoldChange, na.rm = TRUE)) %>%
  ungroup()

# Order the unique genes by log2FoldChange in descending order
unique_genes_ordered_3 <- unique_genes_3 %>%
  arrange(desc(log2FoldChange))

# Create a named vector with log2FoldChange values and gene symbols as names
genes_ordered_3 <- setNames(unique_genes_ordered_3$log2FoldChange, unique_genes_ordered_3$symbol)

genes_ordered_3
```
- For the hallmarks pathways gene sets we'll use msigdbr package.
```{r}
# Load hallmark gene sets
hallmarks_3 <- msigdbr(species = "Homo sapiens", category = "H")
hallmarks_list_3 <- split(hallmarks_3$gene_symbol, hallmarks_3$gs_name)
```
- view hallmark data
```{r}
# Get all genes in hallmarks_list
all_genes_in_hallmarks_3 <- unique(unlist(hallmarks_list_3))

# Get gene identifiers in genes_ordered
genes_in_ordered_3 <- names(genes_ordered_3)

# Check overlap
overlap_genes_3 <- intersect(all_genes_in_hallmarks_3, genes_in_ordered_3)

# Summary of the overlap
cat("Number of genes in hallmarks_list:", length(all_genes_in_hallmarks_3), "\n")
cat("Number of genes in genes_ordered:", length(genes_in_ordered_3), "\n")
cat("Number of overlapping genes:", length(overlap_genes_3), "\n")

```
we received 2 genes.

- Using GSEA() as done in class
```{r}
# Run GSEA
set.seed(123)  # Set seed for reproducibility
gsea_results_3 <- fgsea(pathways = hallmarks_list_3, 
                      stats = genes_ordered_3, 
                      minSize = 15, 
                      maxSize = 500)
gsea_results_3
# Convert results to a data frame
gsea_results_df_3 <- as.data.frame(gsea_results_3)
# Filter significant results (e.g., padj < 0.05)
significant_results_3 <- gsea_results_df_3 %>% filter(padj < 0.05)
significant_results_3
```
we received 0 significant pathways for the third case.
Even though this kind of result might indicate on a troublesome analysis
it's not uncommon to get one hallmark for such a small number of genes (2).


## Next we will preform analysis of the intersection of our 3 cases:
INTERSECT COMMON DEGs from all 3 cases
- We need to intersect the significant genes from all 3 cases to retrieve more reliable DEGs
```{r}
filter.intersect <- filter.sgn.genes.1.nona[filter.sgn.genes.1.nona$symbol %in% filter.sgn.genes.2.nona$symbol ,]
filter.intersect <- filter.intersect[filter.intersect$symbol %in% filter.sgn.genes.3.nona$symbol ,]
filter.intersect
```
- visualizing Venn diagram for common DEGs
- prepare the data
```{r}
# Define the lists of genes for each category
genes_case1 <- na.omit(filter.sgn.genes.1.nona$symbol)
genes_case2 <- na.omit(filter.sgn.genes.2.nona$symbol)
genes_case3 <- na.omit(filter.sgn.genes.3.nona$symbol)

# Create the Venn diagram without the main title
venn.plot <- venn.diagram(
  x = list(
    "Case 1" = genes_case1,
    "Case 2" = genes_case2,
    "Case 3" = genes_case3
  ),
  filename = NULL,
  col = c("lightblue", "red", "green"),
  fill = c("darkblue", "pink", "yellow"),
  cat.pos = 0,          # Position the category names outside the diagram
  cat.dist = 0.05       # Distance of category names from the circles
)

```

- Plot the Venn diagram
```{r}
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 2, widths = unit(c(0.6, 0.4), "npc"))))
pushViewport(viewport(layout.pos.col = 1))
grid.draw(venn.plot)
upViewport()

# Add the descriptive references
pushViewport(viewport(layout.pos.col = 2))
grid.text("Case 1: Non-smokers vs \n reformed for <= 15 years", 
          x = 0, y = 0.85, gp = gpar(fontsize = 10, col = "black"), just = "left")
grid.text("Case 2: Current smokers vs \n reformed for <= 15 years", 
          x = 0, y = 0.75, gp = gpar(fontsize = 10, col = "red"), just = "left")
grid.text("Case 3: reformed for > 15 years vs \n reformed for <= 15 years", 
          x = 0, y = 0.65, gp = gpar(fontsize = 10, col = "green"), just = "left")
upViewport()

```
We can see that there are 4 overlapping significent genes acroos all 3 cases!

### GSEA
We'll use functional enrichment analysis with the Hallmark pathways gene sets for the intersected cases.

- First we need to create an ordered vector by the log fold change with the gene 
  symbols as names:
```{r}
# Convert DESeqResults object to a data frame
filter.intersect_df <- as.data.frame(filter.intersect)

# Remove rows with NA in the symbol column
filtered_data <- filter.intersect_df %>%
  filter(!is.na(symbol))

# Average log2FoldChange for duplicate gene symbols
unique_genes <- filtered_data %>%
  group_by(symbol) %>%
  summarize(log2FoldChange = mean(log2FoldChange, na.rm = TRUE)) %>%
  ungroup()

# Order the unique genes by log2FoldChange in descending order
unique_genes_ordered <- unique_genes %>%
  arrange(desc(log2FoldChange))

# Create a named vector with log2FoldChange values and gene symbols as names
genes_ordered <- setNames(unique_genes_ordered$log2FoldChange, unique_genes_ordered$symbol)

genes_ordered
```
- For the hallmarks pathways gene sets we'll use msigdbr package.
```{r}
# Load hallmark gene sets
hallmarks <- msigdbr(species = "Homo sapiens", category = "H")
hallmarks_list <- split(hallmarks$gene_symbol, hallmarks$gs_name)
```
- view hallmarks data 
```{r}
# Summary of the overlap
# Get all genes in hallmarks_list
all_genes_in_hallmarks <- unique(unlist(hallmarks_list))

# Get gene identifiers in genes_ordered
genes_in_ordered <- names(genes_ordered)

# Check overlap
overlap_genes <- intersect(all_genes_in_hallmarks, genes_in_ordered)

# Summary of the overlap
cat("Number of genes in hallmarks_list:", length(all_genes_in_hallmarks), "\n")
cat("Number of genes in genes_ordered:", length(genes_in_ordered), "\n")
cat("Number of overlapping genes:", length(overlap_genes), "\n")

```
We got 1 overlapping genes.

- Using GSEA() as done in class
```{r}
# Run GSEA
set.seed(123)  # Set seed for reproducibility
gsea_results <- fgsea(pathways = hallmarks_list, 
                      stats = genes_ordered, 
                      minSize = 15, 
                      maxSize = 500)
gsea_results
# Convert results to a data frame
gsea_results_df <- as.data.frame(gsea_results)
# Filter significant results (e.g., padj < 0.05)
significant_results <- gsea_results_df %>% filter(padj < 0.05)
significant_results
```

Unlike HW2, we got here zero Hallmark.
As we interpret the GSEA results, we understand that we might have made some pre-processing decisions / aggregating methods / different computation of COMMON DEGs or computation of DEGs for the 3 cases that caused this result.
In addition, though this kind of result might indicate on a troublesome analysis
it's not uncommon to get one hallmark for such a small number of genes (1).

Let us run GSEA for all cases combined.

- First we need to unite the significant genes from all 3 cases to retrieve more reliable DEGs

```{r}
filter.unite <- filter.sgn.genes.1.nona[filter.sgn.genes.1.nona$symbol %in% filter.sgn.genes.2.nona$symbol ,]
filter.unite <- filter.unite[filter.unite$symbol %in% filter.sgn.genes.3.nona$symbol ,]

# Combine the filtered data frames into one (optional)
filter.unite <- rbind(filter.sgn.genes.1.nona, filter.sgn.genes.2.nona, filter.sgn.genes.3.nona)

# Remove duplicates if necessary
filter.unite <- filter.unite[!duplicated(filter.unite$symbol),]
filter.unite
```
- Next we need to create an ordered vector by the log fold change with the gene 
  symbols as names:
```{r}

# Convert DESeqResults object to a data frame
filter.unite_df <- as.data.frame(filter.unite)

# Remove rows with NA in the symbol column
filtered_data_u <- filter.unite_df %>%
  filter(!is.na(symbol))

# Average log2FoldChange for duplicate gene symbols
unique_genes_u <- filtered_data_u %>%
  group_by(symbol) %>%
  summarize(log2FoldChange = mean(log2FoldChange, na.rm = TRUE)) %>%
  ungroup()

# Order the unique genes by log2FoldChange in descending order
unique_genes_ordered_u <- unique_genes_u %>%
  arrange(desc(log2FoldChange))

# Create a named vector with log2FoldChange values and gene symbols as names
genes_ordered_u <- setNames(unique_genes_ordered_u$log2FoldChange, unique_genes_ordered_u$symbol)

genes_ordered_u
```
- For the hallmarks pathways gene sets we'll use msigdbr package.
```{r}
# Load hallmark gene sets
hallmarks_u <- msigdbr(species = "Homo sapiens", category = "H")
hallmarks_list_u <- split(hallmarks_u$gene_symbol, hallmarks_u$gs_name)
```
- view hallmark data 
```{r}
# Get all genes in hallmarks_list
all_genes_in_hallmarks_u <- unique(unlist(hallmarks_list_u))

# Get gene identifiers in genes_ordered
genes_in_ordered_u <- names(genes_ordered_u)

# Check overlap
overlap_genes_u <- intersect(all_genes_in_hallmarks_u, genes_in_ordered_u)

# Summary of the overlap
cat("Number of genes in hallmarks_list:", length(all_genes_in_hallmarks_u), "\n")
cat("Number of genes in genes_ordered:", length(genes_in_ordered_u), "\n")
cat("Number of overlapping genes:", length(overlap_genes_u), "\n")

```
we got 1166  genes.

- Using GSEA() as done in class
```{r}
# Run GSEA
set.seed(123)  # Set seed for reproducibility
gsea_results_u <- fgsea(pathways = hallmarks_list_u, 
                      stats = genes_ordered_u, 
                      minSize = 15, 
                      maxSize = 500)
gsea_results_u
# Convert results to a data frame
gsea_results_df_u <- as.data.frame(gsea_results_u)
# Filter significant results (e.g., padj < 0.05)
significant_results_u <- gsea_results_df_u %>% filter(padj < 0.05)
significant_results_u
```
we received 6 pathways.

- Visualizing the results
```{r}
# Create a dot plot of significant pathways
ggplot(significant_results_u, aes(x = reorder(pathway, NES), y = NES)) +
  geom_point(aes(size = -log10(padj), color = NES)) +
  coord_flip() +
  labs(x = "Pathway", y = "Normalized Enrichment Score (NES)", 
       title = "GSEA Results", size = "-log10(padj)", color = "NES") +
  theme_minimal()

```
Results and conclusion:
___________________________________

___________________
# Survival Analysis
___________________
- Download CPTAC-3's clinical data 
```{r}
nsclc.clin <- GDCquery_clinic("CPTAC-3", "clinical")
#View(gbm.clin)
```

- Take only "Lung" cancer samples
```{r}
tissue <- c( "Lower lobe, lung", "Lung, NOS", "Middle lobe, lung", "Upper lobe, lung")

filtered_data <- nsclc.clin[nsclc.clin$tissue_or_organ_of_origin %in% tissue, ]
nsclc.clin <- filtered_data
```

- Choose only information that is relevant to survival analysis
```{r}
nsclc.clin.df = nsclc.clin[, 
                           c("submitter_id",
                             "gender",
                             "vital_status", 
                             "cause_of_death",
                             "ajcc_pathologic_stage", 
                             "days_to_death",
                             "days_to_last_follow_up",
                             "tobacco_smoking_status",
                             "cigarettes_per_day")]
```

- Before we can proceed, we need to change part of this information in a way that is acceptable to the methods from the survival package we are using:
```{r}
# create a new boolean variable that has TRUE for dead patients
# and FALSE for live patients
nsclc.clin.df$deceased = nsclc.clin.df$vital_status == "Dead"

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to days_to_last_follow_up for patients who
# are still alive
nsclc.clin.df$overall_survival = ifelse(nsclc.clin.df$deceased,
                                   nsclc.clin.df$days_to_death,
                                   nsclc.clin.df$days_to_last_follow_up)

# convert overall_survival to numeric for Surv object
nsclc.clin.df$overall_survival <- as.numeric(as.character(nsclc.clin.df$overall_survival))
```
- Form a dependent variable out of the overall_survival and deceased information:
```{r}
lung.Surv <- Surv(nsclc.clin.df$overall_survival, nsclc.clin.df$deceased)
```

## Kaplan-Meier plots

### First start with basic question: does the amount of cigarettes per day influence survival in lung cancer patients?
- For doing so, we will divide the patients into 3 groups: Patients who smoke over 20 cigarettes per day, patients who smoke under 20 cigarettes per day,
and patients who don't smoke at all. 

```{r}
nsclc.clin.df <- nsclc.clin.df %>%
  mutate(cigarettes_per_day = case_when(
    tobacco_smoking_status == "Lifelong Non-Smoker" ~ 0,
    TRUE ~ cigarettes_per_day  # Keep the original value otherwise
  ))

nsclc.clin.df <- nsclc.clin.df %>%
  mutate(ccc = case_when(
    cigarettes_per_day == 0   ~ "none",
    cigarettes_per_day <  20  ~ "less then 20",
    cigarettes_per_day >= 20  ~ "more then 20",
    is.na(cigarettes_per_day) ~ NA_character_
  ))
table(nsclc.clin.df$ccc, useNA = "always")
```

```{r}
ggsurvplot(
  fit=survfit(lung.Surv ~ ccc, data=nsclc.clin.df),
  data=nsclc.clin.df,
  xlab = "Days",
  legend.title = "cigarettes per day",
  legend.labs = c("less then 20", "more then 20", "none"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.3,
  ggtheme = theme_bw())
```
### Lets continue with more interesting question: how does tumor stage affect survival?

The tumor_stage variable that TCGA provides for this tumor contains both stages and sub-stages, eg stage IIIA or stage IVB. We want to join together the sub-stages, to increase the group size and reduce complexity (and thus increase the power of the logrank statistics).

```{r}
# remove any of the numbers "1", "2" or "3", but only if they are at the end of the name
# eg "Stage IA2" would become simply "Stage IA"
nsclc.clin.df$ajcc_pathologic_stage = gsub("[123]$", "", nsclc.clin.df$ajcc_pathologic_stage)

# remove any of the letters "A", "B" or "C", but only if they are at the end of the name
# eg "stage IIIA" would become simply "stage III"
nsclc.clin.df$ajcc_pathologic_stage = gsub("[ABC]$", "", nsclc.clin.df$ajcc_pathologic_stage)

# we remove those with stage "not reported", since they are unknown
nsclc.clin.df[which(nsclc.clin.df$ajcc_pathologic_stage == "Not Reported"), "ajcc_pathologic_stage"] = NA

# finally, we also remove those with tumor stage 4, since they are too few
nsclc.clin.df[which(nsclc.clin.df$ajcc_pathologic_stage == "Stage IV"), "ajcc_pathologic_stage"] = NA

table(nsclc.clin.df$ajcc_pathologic_stage)
```
```{r}
ggsurvplot(
  fit=survfit(lung.Surv ~ ajcc_pathologic_stage, data=nsclc.clin.df),
  data=nsclc.clin.df,
  legend.title = "tumor stage",
  legend.labs = c("Stage I", "Stage II", "Stage III"),
  xlab = "Days",
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.35,
  ggtheme = theme_bw())
```
We can see that the p-value we get from this analysis is 0.0083 that is lower than 0.05.
By this results, we can conclude that pathological stage affect the survival time of individuals.
The curves show that the higher the pathological stage, the lower the chance of survival.

### To end this part, we ask one last quesion: how does smoking status affect survival?

```{r}
table(nsclc.clin.df$tobacco_smoking_status)
```
```{r}
# we remove those with Smoking history not documented
nsclc.clin.df[which(nsclc.clin.df$tobacco_smoking_status == "Smoking history not documented"), "tobacco_smoking_status"] = NA

# we also remove current reformed smokers, that have no information about the period of time they quit smoking
nsclc.clin.df[which(nsclc.clin.df$tobacco_smoking_status == "Current Reformed Smoker, Duration Not Specified"), "tobacco_smoking_status"] = NA
```

```{r, fig.width=10, fig.height=5}
ggsurvplot(
  fit=survfit(lung.Surv ~ tobacco_smoking_status, data=nsclc.clin.df),
  data=nsclc.clin.df,
  xlab = "Days",
  legend.title = "Smoking Status",
  legend.labs = c("quit less then 5 years", "quit more then 5 years", "Current Smoker", "Lifelong Non-Smoker"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.3,
  ggtheme = theme_bw()
  )
```
______________________________
# Gene expression and survival
______________________________

For convenience we will create a function that return significant genes from previews analyses
```{r}
#' Read DEseq results from csv
#' @param csv_file The significant gens 
#' @param num_genes The number of gens to read.
#' @return The first num_genes as vecor
#' --------------------------------------------------------------------------
read_deseq_results <- function(csv_file, num_gens) {
  
  lines <- readLines(csv_file)
  gene_list <- c()
  
  for (i in 2:min(num_gens, length(lines))) {
    
    line <- lines[i]
    fields <- strsplit(line, ",")[[1]]
    
    # Extract the Gene name: e.g. "ENSG00000007171.18"
    first_field <- gsub("\"", "", fields[1])
    
    gene_list <- c(gene_list, first_field[1])
  }
  
  return(gene_list)
}
```

### First expeirement
### Survival Analysis on groups: Lifelong Non-Smoker VS Current Reformed Smoker for < or = 15 yrs
```{r}
group1 = "Lifelong Non-Smoker"
group2 = "Current Reformed Smoker for < or = 15 yrs"
```

#### Normal tissue samples
```{r}
dds <-nsclc.normal.dds
data<-nsclc.data.normal
```

top 30 candidates
```{r}
gene_list <- read_deseq_results(csv_file = "signif_results_normal_1.csv", num_gens = 30)
gene_list
```
we will using the log-rank test for searching genes that have strong connection to Survival (pvalue < 0.05)   

```{r}
#A DESeq2 function that can let us extract the normalized values of gene:
d <- plotCounts(dds, gene="ENSG00000232973.13", intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c(group1, group2)
d <- d[d$tobacco_smoking_status %in% selected_statuses, ]


#Keep sample names for those with top and bottom median of gene for later
median.1 <- median(d[d$tobacco_smoking_status == group1,]$count)
median.2 <- median(d[d$tobacco_smoking_status == group2,]$count)
smoking_status = ifelse(median.1 < median.2, group1, group2)
median = ifelse(median.1 < median.2, median.1,median.2)

low <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count <= median,])
high <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count >= median,])

# Take only gene high/low samples 
nsclc.clin.gene <- nsclc.clin.df[nsclc.clin.df$submitter_id %in% c(data@colData[high,]$submitter_id,
                                                     data@colData[low,]$submitter_id),]


# Add a column that indicates which samples is high/low
nsclc.clin.gene$expression <- ifelse(nsclc.clin.gene$submitter_id %in% data@colData[high,]$submitter_id, "high", "low")

# Create a survival plot with Log-rank test
fit = survfit(Surv(overall_survival, deceased) ~ expression, data=nsclc.clin.gene)
results_df <-surv_pvalue(fit)
pval <- results_df$pval
pval
```
we found a significant gene ENSG00000232973.13 (CYP1B1-AS1)

Lets see the difference between the counts:  
```{r}
ggplot(d[d$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "CYP1B1-AS1 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
And Plot a survival curve: 
```{r, fig.width=6, fig.height=6}
ggsurvplot(fit,
  data=nsclc.clin.gene,
  title = "CYP1B1-AS1",
  xlab = "Time since diagnosis (Days)",
  legend.title = "expression",
  legend.labs = c("above median", "below median"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.2,
  ggtheme = theme_bw())
```
#### Primary Tumor samples
```{r}
dds <-nsclc.tumor.dds
data<-nsclc.data.tumor
```

top 30 candidates
```{r}
gene_list <- read_deseq_results(csv_file = "signif_results_tumor_1.csv", num_gens = 30)
gene_list
```
we will using the log-rank test for searching genes that have strong connection to Survival (pvalue < 0.05)   

```{r}
#A DESeq2 function that can let us extract the normalized values of gene:
d <- plotCounts(dds, gene="ENSG00000007171.18", intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c(group1, group2)
d <- d[d$tobacco_smoking_status %in% selected_statuses, ]


#Keep sample names for those with top and bottom median of gene for later
median.1 <- median(d[d$tobacco_smoking_status == group1,]$count)
median.2 <- median(d[d$tobacco_smoking_status == group2,]$count)
smoking_status = ifelse(median.1 < median.2, group1, group2)
median = ifelse(median.1 < median.2, median.1,median.2)

low <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count <= median,])
high <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count >= median,])

# Take only gene high/low samples 
nsclc.clin.gene <- nsclc.clin.df[nsclc.clin.df$submitter_id %in% c(data@colData[high,]$submitter_id,
                                                     data@colData[low,]$submitter_id),]


# Add a column that indicates which samples is high/low
nsclc.clin.gene$expression <- ifelse(nsclc.clin.gene$submitter_id %in% data@colData[high,]$submitter_id, "high", "low")

# Create a survival plot with Log-rank test
fit = survfit(Surv(overall_survival, deceased) ~ expression, data=nsclc.clin.gene)
results_df <-surv_pvalue(fit)
pval <- results_df$pval
pval
```
we found a significant gene ENSG00000007171.18 (NOS2)

Lets see the difference between the counts:  
```{r}
ggplot(d[d$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "NOS2 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
And Plot a survival curve: 
```{r, fig.width=6, fig.height=6}
ggsurvplot(fit,
  data=nsclc.clin.gene,
  title = "NOS2",
  xlab = "Time since diagnosis (Days)",
  legend.title = "expression",
  legend.labs = c("above median", "below median"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.2,
  ggtheme = theme_bw())
```
### Now, lets move on to our second expeirement
### Survival Analysis on groups: Current Smoker VS Current Reformed Smoker for < or = 15 yrs
```{r}
group1 = "Current Smoker"
group2 = "Current Reformed Smoker for < or = 15 yrs"
```

#### well start like before with Normal tissue samples
```{r}
dds <-nsclc.normal.dds
data<-nsclc.data.normal
```

top 30 candidates
```{r}
gene_list <- read_deseq_results(csv_file = "signif_results_normal_2.csv", num_gens = 30)
gene_list
```
we will using the log-rank test for searching genes that have strong connection to Survival (pvalue < 0.05)   

```{r}
#A DESeq2 function that can let us extract the normalized values of gene:
d <- plotCounts(dds, gene="ENSG00000197696.10", intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c(group1, group2)
d <- d[d$tobacco_smoking_status %in% selected_statuses, ]


#Keep sample names for those with top and bottom median of gene for later
median.1 <- median(d[d$tobacco_smoking_status == group1,]$count)
median.2 <- median(d[d$tobacco_smoking_status == group2,]$count)
smoking_status = ifelse(median.1 < median.2, group1, group2)
median = ifelse(median.1 < median.2, median.1,median.2)

low <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count <= median,])
high <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count >= median,])

# Take only gene high/low samples 
nsclc.clin.gene <- nsclc.clin.df[nsclc.clin.df$submitter_id %in% c(data@colData[high,]$submitter_id,
                                                     data@colData[low,]$submitter_id),]


# Add a column that indicates which samples is high/low
nsclc.clin.gene$expression <- ifelse(nsclc.clin.gene$submitter_id %in% data@colData[high,]$submitter_id, "high", "low")

# Create a survival plot with Log-rank test
fit = survfit(Surv(overall_survival, deceased) ~ expression, data=nsclc.clin.gene)
results_df <-surv_pvalue(fit)
pval <- results_df$pval
pval
```
we found a significant gene ENSG00000197696.10 (NMB)

Lets see the difference between the counts:  
```{r}
ggplot(d[d$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "NMB Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
And Plot a survival curve: 
```{r, fig.width=6, fig.height=6}
ggsurvplot(fit,
  data=nsclc.clin.gene,
  title = "NMB",
  xlab = "Time since diagnosis (Days)",
  legend.title = "expression",
  legend.labs = c("above median", "below median"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.2,
  ggtheme = theme_bw())
```
#### Primary Tumor samples
```{r}
dds <-nsclc.tumor.dds
data<-nsclc.data.tumor
```

top 30 candidates
```{r}
gene_list <- read_deseq_results(csv_file = "signif_results_tumor_2.csv", num_gens = 30)
gene_list
```
we will using the log-rank test for searching genes that have strong connection to Survival (pvalue < 0.05)   

```{r}
#A DESeq2 function that can let us extract the normalized values of gene:
d <- plotCounts(dds, gene="ENSG00000105825.14", intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c(group1, group2)
d <- d[d$tobacco_smoking_status %in% selected_statuses, ]


#Keep sample names for those with top and bottom median of gene for later
median.1 <- median(d[d$tobacco_smoking_status == group1,]$count)
median.2 <- median(d[d$tobacco_smoking_status == group2,]$count)
smoking_status = ifelse(median.1 < median.2, group1, group2)
median = ifelse(median.1 < median.2, median.1,median.2)

low <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count <= median,])
high <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count >= median,])

# Take only gene high/low samples 
nsclc.clin.gene <- nsclc.clin.df[nsclc.clin.df$submitter_id %in% c(data@colData[high,]$submitter_id,
                                                     data@colData[low,]$submitter_id),]


# Add a column that indicates which samples is high/low
nsclc.clin.gene$expression <- ifelse(nsclc.clin.gene$submitter_id %in% data@colData[high,]$submitter_id, "high", "low")

# Create a survival plot with Log-rank test
fit = survfit(Surv(overall_survival, deceased) ~ expression, data=nsclc.clin.gene)
results_df <-surv_pvalue(fit)
pval <- results_df$pval
pval
```
we found a significant gene ENSG00000105825.14 (TFPI2)

Lets see the difference between the counts:  
```{r}
ggplot(d[d$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "TFPI2 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
And Plot a survival curve: 
```{r, fig.width=6, fig.height=6}
ggsurvplot(fit,
  data=nsclc.clin.gene,
  title = "TFPI2",
  xlab = "Time since diagnosis (Days)",
  legend.title = "expression",
  legend.labs = c("above median", "below median"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.2,
  ggtheme = theme_bw())
```
### Finally, last expeirement
### Survival Analysis on groups: Current Reformed Smoker for > 15 yrs, Current Reformed Smoker for < or = 15 yrs
```{r}
group1 = "Current Reformed Smoker for > 15 yrs"
group2 = "Current Reformed Smoker for < or = 15 yrs"
```

#### Normal tissue samples
```{r}
dds <-nsclc.normal.dds
data<-nsclc.data.normal
```

top 30 candidates
```{r}
gene_list <- read_deseq_results(csv_file = "signif_results_normal_3.csv", num_gens = 30)
gene_list
```
we will using the log-rank test for searching genes that have strong connection to Survival (pvalue < 0.05)   

```{r}
#A DESeq2 function that can let us extract the normalized values of gene:
d <- plotCounts(dds, gene="ENSG00000232573.1", intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c(group1, group2)
d <- d[d$tobacco_smoking_status %in% selected_statuses, ]


#Keep sample names for those with top and bottom median of gene for later
median.1 <- median(d[d$tobacco_smoking_status == group1,]$count)
median.2 <- median(d[d$tobacco_smoking_status == group2,]$count)
smoking_status = ifelse(median.1 < median.2, group1, group2)
median = ifelse(median.1 < median.2, median.1,median.2)

low <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count <= median,])
high <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count >= median,])

# Take only gene high/low samples 
nsclc.clin.gene <- nsclc.clin.df[nsclc.clin.df$submitter_id %in% c(data@colData[high,]$submitter_id,
                                                     data@colData[low,]$submitter_id),]


# Add a column that indicates which samples is high/low
nsclc.clin.gene$expression <- ifelse(nsclc.clin.gene$submitter_id %in% data@colData[high,]$submitter_id, "high", "low")

# Create a survival plot with Log-rank test
fit = survfit(Surv(overall_survival, deceased) ~ expression, data=nsclc.clin.gene)
results_df <-surv_pvalue(fit)
pval <- results_df$pval
pval
```
we found a significant gene ENSG00000232573.1 (RPL3P4)

Lets see the difference between the counts:  
```{r}
ggplot(d[d$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "RPL3P4 Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
And Plot a survival curve: 
```{r, fig.width=6, fig.height=6}
ggsurvplot(fit,
  data=nsclc.clin.gene,
  title = "RPL3P4",
  xlab = "Time since diagnosis (Days)",
  legend.title = "expression",
  legend.labs = c("above median", "below median"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.2,
  ggtheme = theme_bw())
```
#### Primary Tumor samples
```{r}
dds <-nsclc.tumor.dds
data<-nsclc.data.tumor
```

top 30 candidates
```{r}
gene_list <- read_deseq_results(csv_file = "signif_results_tumor_3.csv", num_gens = 40)
gene_list
```
we will using the log-rank test for searching genes that have strong connection to Survival (pvalue < 0.05)   

```{r}
#A DESeq2 function that can let us extract the normalized values of gene:
d <- plotCounts(dds, gene="ENSG00000232680.2", intgroup="tobacco_smoking_status", returnData=TRUE)
selected_statuses <- c(group1, group2)
d <- d[d$tobacco_smoking_status %in% selected_statuses, ]


#Keep sample names for those with top and bottom median of gene for later
median.1 <- median(d[d$tobacco_smoking_status == group1,]$count)
median.2 <- median(d[d$tobacco_smoking_status == group2,]$count)
smoking_status = ifelse(median.1 < median.2, group1, group2)
median = ifelse(median.1 < median.2, median.1,median.2)

low <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count <= median,])
high <- rownames(d[d$tobacco_smoking_status == smoking_status & d$count >= median,])

# Take only gene high/low samples 
nsclc.clin.gene <- nsclc.clin.df[nsclc.clin.df$submitter_id %in% c(data@colData[high,]$submitter_id,
                                                     data@colData[low,]$submitter_id),]


# Add a column that indicates which samples is high/low
nsclc.clin.gene$expression <- ifelse(nsclc.clin.gene$submitter_id %in% data@colData[high,]$submitter_id, "high", "low")

# Create a survival plot with Log-rank test
fit = survfit(Surv(overall_survival, deceased) ~ expression, data=nsclc.clin.gene)
results_df <-surv_pvalue(fit)
pval <- results_df$pval
pval
```
we found a significant gene ENSG00000232680.2 (AC002511.3 )

Lets see the difference between the counts:  
```{r}
ggplot(d[d$count < 200,], aes(tobacco_smoking_status, count)) + 
  geom_boxplot(aes(fill=tobacco_smoking_status)) + 
  labs(title = "AC002511.3  Expression by Smoking Status",
       x = "Tobacco Smoking Status",
       y = "Count",
       fill = "Smoking Status",
       color = "Smoking Status") +  # Add labels for axes and legend
  theme_minimal() +  # Use a minimal theme
  theme(axis.text.x = element_text(size = 8),  # Rotate x-axis labels
        plot.title = element_text(hjust = 0.5))  # Center the plot title
```
And Plot a survival curve: 
```{r, fig.width=6, fig.height=6}
ggsurvplot(fit,
  data=nsclc.clin.gene,
  title = "AC002511.3 ",
  xlab = "Time since diagnosis (Days)",
  legend.title = "expression",
  legend.labs = c("above median", "below median"),
  pval=T,
  risk.table=T,
  risk.table.col="strata",
  tables.height = 0.2,
  ggtheme = theme_bw())
```
  ## Primary Tumor  
  ----Add your code-----
