---
title: "Homework - 3"
author: "Almog Angel"
students: "<Amit Gabay> <206040768> <Hen Ori Israel> <308215268>"
date: "07/03/2024"
output: html_document
---

In this homework assignment, we will analyze genotyping data of 9 dog breeds.

Go to the study "The Shepherdsâ€™ Tale: A Genome-Wide Study across 9 Dog Breeds Implicates Two Loci in the Regulation of Fructosamine Serum Concentration in Belgian Shepherds" by Forsberg et al - https://doi.org/10.1371/journal.pone.0123173

- Read the abstract, introduction, discussion and the GWAS results section and answer the following questions in brief:
(1) What is Fructosamine? and what disease is it associated with?
(2) Which dog breeds are at low risk of developing diabetes?
(3) What do the authors aim to find in this study?
(4) Why are domestic dogs useful models for genetic studies of human complex diseases?


```{}
# Write yourt answers here:
(1) 
Fructosamine is a compound that is created from glycation reactions between a sugar and a protein.
Fructosamine is a major biomarker of glycaemia, where serum concentrations demonstrate the glucose cycle over a period of past few weeks to months.
The disease associated with Fructosamine is diabetes mellitus.

(2)
According to the study, the dog breeds that are at low risk of developing diabetes mellitus disease are Belgian shepherds, German shepherds and Boxers.
(In the introduction part it is hinted that Golden retrievers may also be at low risk but no further studies have been shown)

(3)
The authors searched for genetic factors that influence variation in serum fructosamine concentration in healthy dogs. 
A more general goal of the study is to better understand the regulation of glucose metabolism, which may improve diagnostic and therapeutic methods for diabetes mellitus.

(4)
Domestic dogs are useful models for genetic studies of human complex diseases due to several reasons - modern dogs and humans share several common and complex diseases. 
This fits the fact that domestic dogs have been living alongside humans for thousands of years, which indicated that the two shared the same environment for many generations. 
In addition, strong selection for certain traits, that have occured for the last 200 years, created dog breeds with unique diversity among species of mammals. 
Nevertheless, while dog breeds were formed over time, a limitation of genetic variation within breeds had increased, leading to enrichment of some risk alleles. 
This made specific diseases more prevalent in domestic dogs. 
The strong recent selection caused canine genomes to be characterised by relatively long haplotypes.
```

- Load packages
```{r}
library(statgenGWAS)
library(factoextra)
library(ggplot2)
```

- Set working directory
```{r}
setwd("C:/Users/Amit_g/Desktop/Bio_HW/HW3")
```

- Load data and take a look:
```{r}
geno <- read.table("dogs.geno") # SNPs matrix of samples from nine different dog breeds
map <- read.table("dogs.map") # Map table with the SNPs ID, chromosome and position
pheno <- read.table("dogs.pheno", header = T) # Phenotypic and metadata for each dog sample (genotype)
View(geno)
View(map)
View(pheno)
```

# Part 1: Bobby and population genetics -----

Last month you adopted a 2.5 years old mixed-breed dog named Bobby.
He is very cute, friendly and quite big (weighs 30kg).
Therefore, you are pretty confident he is a mix of big dog breeds.

To find out what dog breed Bobby most likely related to, you did what any reasonable person would do - asked your friend from the faculty of biology to genotype a sample of Bobby.

For part 1, we would like to analyze only "big dogs" (over 20kg) as candidates for their relation to Bobby.

- Use the data in `pheno` to make a variable called `candidate_breeds` that holds a vector of dog breed names
(1) Make a boxplot of dogs breeds vs. weight*
(2) Based on the boxplot, use only dog breeds that reach 20kg and higher as candidates
* Color, order, use axis labels and remove NAs (pretty graphs get extra points!)
Check this website for examples - https://www.r-graph-gallery.com/
```{r}
##### 1 ##### 
pheno <- pheno[!is.na(pheno$Breed) | !is.na(pheno$Body_weight),]
ggplot(pheno, aes(Breed, Body_weight)) + theme(axis.text.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  geom_boxplot(aes(fill=Breed)) + ggtitle("Dogs breeds vs. weight")
```


```{r}
##### 2 ##### 
candidate_breeds <- unique(pheno$Breed)
candidate_breeds <- candidate_breeds[! candidate_breeds == "Dachshund" & ! candidate_breeds == "Cavalier King Charles Spaniel"]
```

- Think of a computational method that we learned in class that will help you visualize and decide what dog breed Bobby is most likely related to.
(1) Use only data from `geno` that belongs to `candidate_breeds`
(2) Load Bobby's genotyping results file: `bobby.geno`
(3) Do not forget to plot (pretty graphs get extra points!)
TIP: use one of `factoextra` functions for the visualization and color the different dogs breeds
```{r}
##### 1 ##### 
dogs.candidates.ids <- rownames(pheno[pheno$Breed %in% candidate_breeds, ])
geno.candidate_breeds <- geno[dogs.candidates.ids, ]

##### 2 #####
bobby.geno <- read.table("bobby.geno")
View(bobby.geno)

##### 3 #####
kmeans.out <- kmeans(geno.candidate_breeds, centers = 7)
fviz_cluster(kmeans.out, data = geno.candidate_breeds, geom = "point") +
  theme(plot.title = element_text(hjust = 0.5))
```
```{r}
# calc the euclidean dist between bobby and the other
mat <- rbind(kmeans.out$centers, bobby.geno)
centers_dist <- dist(mat, diag = TRUE, upper = TRUE, method = "euclidean")
dist_mat <- as.matrix(centers_dist)
```
```{r}
# find out what bobby is
bobby_cluster_index = which.min(dist_mat["Bobby", 1:7])
unique_clusters <- unique(kmeans.out$cluster)
bobby_cluster_name = candidate_breeds[unique_clusters][bobby_cluster_index]
cat("Bobby is:", bobby_cluster_name, "\n")
```
- Answer the following questions:
(1) Which dog breed is the most similar to Bobby?
(2) Look at the clusters of Labrador, Belgian and German Shepherd and imagine lines that connect the centroids of those cluster which results in an equilateral triangle.
Is it true to say that the similarity between Belgian Shepherd to German Shepherd is equal to the similarity between Belgian Shepherd to Labrador? Explain.
```{}
#answers:
(1) The dog breed that is the most similar to Bobby is Labrador Retriever.
(2)
The say above is not true, since we are observing only a two-dimensional space, spanned by the two vectors with "most variance", where one of them is more significant than the other. That means that the contribution of each axis to the similarity is different. 
Thus, the same euclidian distance of one centroids pair doesn't imply that the similarity of the corresponding clusters will be equal to the similarity of another clusters pair with the same euclidian distance between centroids. 
We also know that there is a significant amount of variance that is caught by other spatial vectors (the PCA eigenvectors), which affects the similarity as well, beyond our limited perception in the 2-D plot.
We'll mention that the term we used - "clusters similarity", refers to rough distance between the whole clusters.
```


# Part 2: GWAS -----

- Make sure geno, map and pheno left unchanged.
```{r}
geno <- read.table("dogs.geno")
map <- read.table("dogs.map")
pheno <- read.table("dogs.pheno", header = T)
```

- Create a gData object and call it `gDataDogs`:
(1) Make sure that your data match the instructions in `?createGData()`
(2) Make a list called `dogsPhenoList` of different dog breeds out of `pheno` and use only the column `genotype` and `FRUCTO`
```{r}
#(1)
?createGData()
colnames(map)[2:3] <- c("chr", "pos")
rownames(map) <- map$V1
map <- map[, c("chr", "pos")]

#(2)
dogsPhenoList <- dplyr::select(pheno, "sex", "FRUCTO")
dogsPhenoList$sex <- rownames(dogsPhenoList)
rownames(dogsPhenoList) <- 1:length(rownames(dogsPhenoList))
names(dogsPhenoList)[1] <- "genotype"
dogsPhenoList <- split(x = dogsPhenoList[c("genotype", "FRUCTO")], 
                        f = pheno[["Breed"]])

gDataDogs <- createGData(geno = geno, map = map, pheno = dogsPhenoList)
```

- Run a GWAS analysis for Belgian Shepherd with fructosamine concentrations:
(1) Show the QQ- and Manhattan plots 
(2) Print significant SNP(s)
```{r}
belgianSheperdGWAS <- runSingleTraitGwas(gData = gDataDogs,
                              trials = "Belgian Shepherd",
                              traits = c("FRUCTO"))
#(1)
plot(belgianSheperdGWAS, plotType = "qq", trait = "FRUCTO")
plot(belgianSheperdGWAS, plotType = "manhattan", trait = "FRUCTO")

#(2)
print((belgianSheperdGWAS$signSnp$"Belgian Shepherd"), row.names = FALSE)

```

- Answer the following questions using the QQ- and Manhattan plot:
(1) While presenting your GWAS results in an international dog-lovers confrence, an elderly woman with a well-groomed Pekingese dog in her bag, challenge your interpretation from the QQ plot. She claims that the deviation observed in the QQ plot could be due to difference in the population structure rather than true genetic associations. Briefly describe how you would respond to her concerns given your QQ plot results.
(2) A curious breeder from the audience, intrigued by the implications, posed the another question: "Could you explain how the location of these significant SNP(s) on your plot help us understand genetic associations with the trait?"
```{}
# Write yourt answers here:

```


- Make a boxplot of Fructosamine concentrations for different alleles in Belgian and German Shepherds:
(1) Use only the genotypes of Labrador Retriever, Belgian and German Shepherd
(2) The X-axis should be the different alleles (0, 1, and 2) of the most significant SNP from the Belgian Shepherd GWAS results
(3) Remove rows with NAs in fructosamine concentrations
```{r}

```

- Answer the following questions:
(1) Describe the results from the box-plot with respect the three allels (0, 1 and 2) and the three dog breeds.
(2) Name the most significant SNP, indicates the chromosome and positions.
```{}
# Write your answers here:

```

- Go to - "http://genome-euro.ucsc.edu/cgi-bin/hgGateway".
- Select "Dog" from the list of species in the left.
- Select dog assembly: "May 2005 (Broad/canFam2)".
- Select and use the chromosome and position of the SNP from the first question and click "GO".
- Click on the last layer in the genome browser (Simple Nucleotide Polymorphism - rs23514694)

* The first part of this video can be useful: https://www.youtube.com/watch?v=8U5NhHofPI0

(1) Write the nucleotides combinations of the three different alleles
```{}
# Write your answers here:

```

- Go back to the genome browser
- Zoom out (x100) three times - in the top right.
- Identify human proteins that are mapped using tBLAST to this dog genome region

(2) Choose 5 proteins and check if they are mentioned in the paper. If they are mentioned, explain how they are relevant to the trait.
```{}
# Write your answers here:


```

- I asked ChatGPT to tell a joke about dog breeds:

ChatGPT:
"Why did the Dalmatian go to the eye doctor?
Because he kept seeing spots everywhere!"


- Answer the following questions:
(1) How funny do you find this joke?
(2) If you have a better one, let me know so I will enjoy reading your homework :) - Almog
```{}
# Write your answers here:


```

